KEP-2170: Kubeflow Training V2 API (#2171)
kubeflow/training-operator: docs/proposals/2170-kubeflow-training-v2/README.md

# KEP-2170: Kubeflow Training V2 API

## Authors

- Andrey Velichkevich - [@andreyvelich](https://github.com/andreyvelich)
- Yuki Iwai - [@tenzen-y](https://github.com/tenzen-y)

Creation date: 2024-07-16

Google doc: https://bit.ly/3WzjTlw

## Overview

This document discusses the new Kubeflow Training V2 API.

When we built the
[Kubeflow Training Operator a couple of years ago](https://docs.google.com/document/d/1x1JPDQfDMIbnoQRftDH1IzGU0qvHGSU4W6Jl4rJLPhI/edit?usp=sharing),
Kubernetes lacked better features to support distributed machine learning (ML) training, such as
SuccessPolicy and RestartPolicy (FailurePolicy). Recently, the Kubernetes community launched the
working group Batch, and then the working group actively worked on evolving the batch/v1 `Job` API

> and built [a new `JobSet`](https://github.com/kubernetes-sigs/jobset) API to manage groups of `Jobs`.

Author: Sure, let me add i.

the maintenance costs will continue to increase.

It would be better to replace infrastructure layers with `JobSet` to avoid re-inventing the wheel
and improve the Training Operator

Additionally, introducing `JobSet` as an infrastructure layer would allow us to introduce batch
workload features such as
[the PodFailurePolicy](https://kubernetes.io/docs/concepts/workloads/controllers/job/#pod-failure-policy)
and [the PodDisruptionCondition](https://kubernetes.io/docs/concepts/workloads/controllers/job/#handling-pod-and-container-failures)
easily.

Please also see the [Kubernetes JobSet and Kubeflow Training Operator collaboration document](https://docs.google.com/document/d/1C2ev7yRbnMTlQWbQCfX7BCCHcLIAGW2MP9f7YeKl2Ck/edit?usp=sharing).

### User Value

In addition to the above motivation, we will address the following user feedback while implementation:

- Confusion around Workers: https://github.com/kubeflow/training-operator/issues/1790
- Support batch/v1 `Job` features: https://github.com/kubeflow/training-operator/issues/1718
- ExitCodes for PodFailurePolicy: https://github.com/kubeflow/training-operator/issues/1749

> - Migrate to MPI V2 API: https://github.com/kubeflow/training-operator/issues/1906

Reviewer: There will be three operators for MPI actually.

V1 (which is being deprecated)
V2 (one that uses indexed jobs)
TrainJob.

KubeFlow is hoping to drop V1 soon but not sure when that will happen.

> - Migrate to MPI V2 API: https://github.com/kubeflow/training-operator/issues/1906

Author: We will certainly deprecated MPI V1 Operator as part of this effort.

The question about keeping MPI Operator separate, we need to discuss if we really need to keep it. It is hard to maintain two similar operators.

- We would like to hear from HPC and MPI Operator users if the blueprint idea might be useful for them as well?
- Additionally, If Training Operator can be used for all MPI jobs (ML + HPC) what is the reason to keep MPI Operator separate ?

cc @alculquicondor @tenzen-y

> - Migrate to MPI V2 API: https://github.com/kubeflow/training-operator/issues/1906

Reviewer 2: MPI is more general purpose instead of focusing on training so IMO it makes more sense to keep it separate.

> - Migrate to MPI V2 API: https://github.com/kubeflow/training-operator/issues/1906

Author: I am not sure if we have enough bandwidth to support MPI Operator and Training Operator separately given that all functionality of MPI Operator will be moved to Training Operator.
I would love to hear some objections from HPC users for this proposal, especially if the idea of re-usable MPI runtimes is not applicable for their use-cases.
cc @kubeflow/wg-training-leads

> - Migrate to MPI V2 API: https://github.com/kubeflow/training-operator/issues/1906

Reviewer 3: It's hard to make a call (and objection) before the new interfaces exist. An objection would come if functionality is lost by way of moving everything here without considering non-ML use cases of MPI.

> - Migrate to MPI V2 API: https://github.com/kubeflow/training-operator/issues/1906

Author: That makes sense, so maybe after initial implementation we can have this discussion again.

> - Migrate to MPI V2 API: https://github.com/kubeflow/training-operator/issues/1906

Reviewer 4: IMO, MPI V1 is out definitely. But MPI style launcher mode should be support since DeepSpeed is popular in LLM.

> - Migrate to MPI V2 API: https://github.com/kubeflow/training-operator/issues/1906

Reviewer: In an ideal world, I'd love to have the HPC community aiding/helping with the development of a general MPI operator for Kubernetes that lives outside of Kubeflow. There was some discussion about "moving" MPIOperator to Kubernetes at some point. It sorta died because Kubeflow is part of CNCF but I still see value in building a community supported operator for MPI. 

I think having a limited scope approach for MPI for Kubeflow makes sense. But having MPI under the Kubeflow community has always seemed to confuse HPC/Researchers as I noticed that they don't consider using the MPI Operator from Kubeflow if they are trying to run MPI jobs for fluid dynamics (for example) as Kubeflow is designed for ML usecases.

> - Migrate to MPI V2 API: https://github.com/kubeflow/training-operator/issues/1906

Author: Why HPC community can't contribute to the MPI on Kubernetes with Kubeflow community ? Kubeflow is under CNCF and I don't see any issues for HPC community to work with us to build various HPC runtimes.

>I noticed that they don't consider using the MPI Operator from Kubeflow if they are trying to run MPI jobs for fluid dynamics (for example) as Kubeflow is designed for ML usecases.

I feel that HPC and ML worlds are strongly connected. That is why many folks are looking for HPC technologies to train their models today (e.g. Slurm, MPI, Flux). @vsoch let me know if I am wrong here.

It would be beneficial for both communities if we make Kubernetes more accessible for HPC or AI/ML use-cases.

> - Migrate to MPI V2 API: https://github.com/kubeflow/training-operator/issues/1906

Reviewer 3: You are entirely correct, and I do advocate that we work together! My organization explicitly has permission to contriute to CNCF, so I can continue to offer my time and attention.

> - Migrate to MPI V2 API: https://github.com/kubeflow/training-operator/issues/1906

Reviewer 3: > But having MPI under the Kubeflow community has always seemed to confuse HPC/Researchers as I noticed that they don't consider using the MPI Operator from Kubeflow if they are trying to run MPI jobs for fluid dynamics (for example) as Kubeflow is designed for ML usecases.

What do they use instead? My (personal) observation is that it's hard to setup and get working, but I'm also biased to want to use the Flux Operator when we bring MPI into Kubernetes. If you like we can have a discussion about what could be worked on.

> - Migrate to MPI V2 API: https://github.com/kubeflow/training-operator/issues/1906

Reviewer: > Why HPC community can't contribute to the MPI on Kubernetes with Kubeflow community ? Kubeflow is under CNCF and I don't see any issues for HPC community to work with us to build various HPC runtimes.

They can and should! I think having Kubeflow in CNCF will go a long way in getting more people to contribute. 

I think also some teaching on the fact that people can install MPIOperator without installing Kubeflow platform is a necessary point unfortunately. I have had to explain that a lot more than I'd like to admit.

To make it more general, having examples of scientific computing jobs in this repo would be helpful to explain that this project is not just meant for ML.

I am just unsure which repo/project should the development of a general MPI operator fall under. TrainJob is focused on having a pythonic/simple experience for running distributed jobs. Given how vast/complicated MPI can be I am not sure that we can really have that for a general MPI job at the moment. I think having the kubeflow use case covered here would be sufficient.

> What do they use instead? My (personal) observation is that it's hard to setup and get working, but I'm also biased to want to use the Flux Operator when we bring MPI into Kubernetes. If you like we can have a discussion about what could be worked on.

Either they think its impossible to run MPI and Kubernetes and turn to something like Ray or they keep using something like Slurm without Kubernetes seems to be what I've heard. I haven't really worked in too many places that use MPI to be honest.

> - Migrate to MPI V2 API: https://github.com/kubeflow/training-operator/issues/1906

Author: >I think also some teaching on the fact that people can install MPIOperator without installing Kubeflow platform is a necessary point unfortunately. I have had to explain that a lot more than I'd like to admit.

We know this, as we worked a lot in the last 6 months to explain: What is Kubeflow ? 

- https://www.kubeflow.org/docs/started/introduction/#what-are-standalone-kubeflow-components
- https://www.kubeflow.org/docs/started/installing-kubeflow/#standalone-kubeflow-components

Also, we will remove all content from `kubeflow/kubeflow` repo to avoid confusions: https://github.com/kubeflow/kubeflow/issues/7549

> TrainJob is focused on having a pythonic/simple experience for running distributed jobs. 

The main idea of this proposal is to distinguish blueprint from the actual job that user submits. Since many configuration doesn't need to be changed every job submission (schedulers, run policies, node selectors).

As long as we can create HPC runtime, TrainJob can be utilized to start HPC job (although, I understand the naming confusion).

> - Migrate to MPI V2 API: https://github.com/kubeflow/training-operator/issues/1906

Reviewer 3: > Either they think its impossible to run MPI and Kubernetes and turn to something like Ray or they keep using something like Slurm without Kubernetes seems to be what I've heard. I haven't really worked in too many places that use MPI to be honest.

This sounds like an issue with awareness / advertising. More specifically, the kinds of proxy apps we run in HPC aren't known to the cloud community, so the wires don't cross. I also suspect this - using the "HPC in a box" solutions like HPC Toolkit (Google) or AWS Parallel Cluster instead of Kubernetes solutions.

> - Migrate to MPI V2 API: https://github.com/kubeflow/training-operator/issues/1906

Reviewer 3: > As long as we can create HPC runtime, TrainJob can be utilized to start HPC job (although, I understand the naming confusion).

I agree - the design is analogous. We have a launcher and workers, and the underlying details of how they hook up is specific to the container and commands provided.

> - Migrate to MPI V2 API: https://github.com/kubeflow/training-operator/issues/1906

Reviewer: > As long as we can create HPC runtime, TrainJob can be utilized to start HPC job (although, I understand the naming confusion).

We could call this DistributedJob?

If MPI is included maybe we can have this be a project for running distributed jobs (training is really just a configuration..).

But I won't make that blocking.

> - Migrate to MPI V2 API: https://github.com/kubeflow/training-operator/issues/1906

Reviewer: Either way, I think I am happy with this discussion as is.

In summary, it seems that there is agreement with forcing out v1 MPIOperator. The Training V2 would include MPI Support.

But there is still an open discussion on what do with v2 mpi operator?

> - Migrate to MPI V2 API: https://github.com/kubeflow/training-operator/issues/1906

Author: >We could call this DistributedJob?

`TrainJob` might give us more flexibility, since maybe users just want to use `TrainJob` to Train Model on single Kubernetes worker and export it somewhere.

>But there is still an open discussion on what do with v2 mpi operator?

Yes, let's followup with the community on what we should do with MPI Operator once we have initial implementation.

> - Migrate to MPI V2 API: https://github.com/kubeflow/training-operator/issues/1906

Reviewer: Sounds good to me then!

> - Migrate to MPI V2 API: https://github.com/kubeflow/training-operator/issues/1906

Reviewer 5: Sorry for the lately comment.
I'm thinking that we can provide the MPIJob v3 API as the mpi-operator which has the completely same API as the training-operator TrainJob API. As we discussed previously, I'm wondering if we can reuse the API definition by defining the `mlspec.mpi` API in the mpi-operator side, and then import the `mlspec.mpi` API definition to the training-operator repo.

Anyway, I think that it's nice to have a dedicated KEP for the reusable MPI API definition.

> - Migrate to MPI V2 API: https://github.com/kubeflow/training-operator/issues/1906

Author: Yes, let's followup on this after initial implementation. We need to discuss the use-cases from HPC side when they want to have dedicated controller for them since we support all MPI functionality in Training Operator.

### Personas

We can identify the following personas of Training Operator:

1. **DevOps Engineer**. They are familiar with Kubernetes concepts and they know how to manage the
   Kubernetes workloads. Usually, they are not experts in ML frameworks and ML algorithms.
1. **MLOps Engineer**. They are familiar with ML frameworks and they know how to configure
   distributed PyTorch settings such as rendezvous backends or MPI configuration. Usually, they are
   not experts in Kubernetes and ML algorithms.
1. **Data Scientists**. They create model architectures and advanced ML algorithms to train models.
   They prefer to use Python for their work. They are aware of `torch.nn` APIs, but not with
   `torch.distributed` and Kubernetes concepts to scale model training.

Based on the above personas, we should build an API that everyone will benefit from.

### Goals

> - Introduce the `TrainingRuntime` and `ClusterTrainingRuntime` APIs that will store blueprints

Author: Let's continue discussing it here: https://github.com/kubeflow/training-operator/pull/2171#discussion_r1680412433

### Worker and Node Definition

To better understand what does Nodes and Worker mean in the diagram above,
the following table explains naming that each framework or technology uses:

<table>
  <tr>
   <td><strong>ML Framework or Technology</strong>
   </td>
   <td><strong>Definition of a Single Device (GPU)</strong>
   </td>
   <td><strong>Definition of a Single VM</strong>
   </td>
   <td><strong>Start Command</strong>
   </td>
   <td><strong>Reference Docs</strong>
   </td>
  </tr>
  <tr>
   <td>Kubernetes
   </td>
   <td>Container Resource Unit
   </td>
   <td>Pod’s Container
   </td>
   <td>Any
   </td>
   <td><a href="https://kubernetes.io/docs/concepts/configuration/manage-resources-containers/#resource-units-in-kubernetes">Resource units</a> in K8s
   </td>
  </tr>
  <tr>
   <td>PyTorch
   </td>
   <td>Worker 
<p>
<code>(--nproc-per-node)</code>
   </td>
   <td>Node
<p>
<code>(--nnodes)</code>
   </td>
   <td><code>torchrun</code>
   </td>
   <td><a href="https://pytorch.org/docs/stable/elastic/run.html">PyTorch Elastic</a>
   </td>
  </tr>
  <tr>
   <td>MPI
   </td>
   <td>Slot
<p>
<code>(--np)</code>
   </td>
   <td>Node 
<p>
<code>(--host)</code>
   </td>

>    <td><code>mpirun</code>

Author: Thanks for sharing. Do you have any useful links that we can add here for reference to explain how different MPI implementations map workers and nodes to physical resources ?

>    <td><code>mpirun</code>

Reviewer 3: What I have quickly handy is a guide for resource managers, which more efficiently bootstrap MPI: https://hpc.llnl.gov/banks-jobs/running-jobs/batch-system-cross-reference-guides. I couldn't find a bare bones MPI launcher Rosetta Stone, and I started putting one together tonight for OpenMPI, MPICH, Intel MPI, and spectrum MPI. I want to sanity check a few things but will post the link in the next few days!

>    <td><code>mpirun</code>

Reviewer 3: Alright - here is a quick rundown, for the top set that I see most frequently. https://gist.github.com/vsoch/9ac7c4448dffe656d946edceaa58bd9e

I think to generally say `-n or -np` is the flag for processes is correct, and there is a bit more variation in the host definition (e.g., `-host` vs `-f` vs `-hostfile` for comma separated command line vs file) but for this table it would work to provide an example from one implementation (OpenMPI) as you have done, but then note that the actual interface can vary between implementations, but also within an implementation depending on your needs for topology / how you ask for resources. For example, asking for a total number of nodes (pods) with `-N` and then specifying processes per node, vs providing a hostlist and a _total_ number of processes (and letting MPI figure out the distribution) is different. Asking for a specific topology (which will give you the correct performance) is different. In HPC we typically don't interact with MPI directly, but use a workload launchers to more efficiently bootstrap it. As long as the message isn't strongly "there is one MPI, it's mpirun, and this is what it looks like" I think it's OK.

>    <td><code>mpirun</code>

Author: Thanks for summarizing this! Let me add this as a reference to better understand the signature between different MPI implementations.

	DatasetConfig *DatasetConfig `json:"datasetConfig,omitempty"`

	// Configuration for the pre-trained model and location for model output
	ModelConfig *ModelConfig `json:"modelConfig,omitempty"`

	// Custom metadata to apply for Job, JobSet, etc.
	Labels      map[string]string `json:"labels,omitempty"`
	Annotations map[string]string `json:"annotations,omitempty"`
}

type TrainingRuntimeRef struct {
	// Name for the training runtime.
	Name string `json:"name"`
	// Namespace for the runtime.
	// If namespace is set, TrainingRuntime is used. Otherwise, ClusterTrainingRuntime is used.
	Namespace string `json:"namespace,omitempty"`
}

type TrainJobStatus struct {
	// Conditions for the TrainJob
	Conditions []metav1.Condition `json:"conditions,omitempty"`
}

```

This table explain rationale for each `TrainJob` parameter:

<table>
  <tr>
   <td><strong>Parameter</strong>
   </td>
   <td><strong>What is it ?</strong>
   </td>
  </tr>
  <tr>
   <td><code>TrainingRuntimeRef</code>
   </td>
   <td>Reference to the existing <code>TrainingRuntime</code> that pre-deployed by platform engineers
   </td>
  </tr>
  <tr>
   <td><code>TrainerConfig</code>
   </td>

>    <td>Configuration for the Trainer such as image, number of nodes, accelerators.

Author: I think, the main goal is to provide minimal set of configurations that user wants to override for every job.
As we discussed before, schedulers and run policies need to be configured via blueprints/runtimes.

  trainerConfig:
    image: docker.io/custom-training
    command:
      - torchrun train.py
    numNodes: 5
    resourcesPerNode:
      requests:
        nvidia.com/gpu: 2
```

The above command will be converted as follows:

```bash
torchrun --nnodes=5 --nproc-per-node=2 train.py
```

Additionally, the Kubeflow Training SDK allows to create the above `TrainJob` using the Python API:

```python
def train_func():
    import torch
    class Net(torch.nn.Module):
        """Create the PyTorch Model"""
        ...
    model = Net()

    # Attach model to the distributor
    torch.distributed.init_process_group(backend="nccl")
    model = torch.nn.parallel.DistributedDataParallel(model)

    # Train model
    model.train()

# Use Kubeflow SDK to create TrainJob.
from kubeflow.training import TrainingClient

TrainingClient().train(
    name="torch-ddp",
    func=train_func,
    num_nodes=5,
    resources_per_node={"gpu": 2},
)
```

### Example of LLM Fine-Tuning

This example shows how to create `TrainJob` to fine-tune LLama 7b:

```yaml
apiVersion: kubeflow.org/v2alpha1
kind: TrainJob
metadata:
  name: tune-llama-with-yelp
  namespace: tenant-alpha
spec:

>   trainingRuntimeRef:

Reviewer 6: Regarding the model saving output, whether it's S3, Google Storage, or another option, it would be beneficial to have a unified solution that can be utilized by Kubeflow, the Training Operator, Pipelines, or any other Kubeflow components. While this isn't a necessity for V2, it's something we should keep in mind for future development.

>   trainingRuntimeRef:

Reviewer 3: Agree here, I mentioned something along these lines in one of my comments.

>   trainingRuntimeRef:

Reviewer 7: +1 to keep flexibility for multiple, pluggable storage proposals

>   trainingRuntimeRef:

Author: > And the path selected by the user will too. For this kind of variable, it can't be a direct part of this training runtime ref (nor can it be hidden so the user doesn't see it and it leads to unexpected behavior).

I think, we should consider that for the fine-tuning user can modify:
- Trainer Hyperparameters (e.g. number of epochs).
- Pre-Trained Model Architecture.
- Dataset location.
- Model location.

As part of my question above, we should discuss how do we want to support LLM runtimes ?
```
What LLM Fine-Tuning Runtimes do we want to add initially. E.g. Llama-7b, Gemma-7b, Llama-70b?
```

For example, if we introduce Llama-70b runtime, user doesn't want to modify the model location. And we also configure set of recipes that are required to fine-tune that model.
I think, we should investigate [the `torchtune` tool](https://github.com/pytorch/torchtune/tree/main), and see if we can re-use it. Since it is natively integrated with `torchrun`.

Additionally, if we introduce more `generic` LLM fine-tuning runtimes, user should be able to configure model, dataset, or trainer with their requirements.

But we are looking for community suggestions on this topic.

    name: torch-tune-llama-7b
  datasetConfig:
    s3:
      bucket: datasets
      path: custom-datasets/yelp-review
```

### The Trainer Config API

The `TrainerConfig` represents the APIs that data scientists can use to configure trainer settings:

> 

Author: Yes, but we also implement Python API to manage TrainJobs. It can be used as in this example: https://www.kubeflow.org/docs/components/training/getting-started/#getting-started-with-pytorchjob.

Additionally, we want to add UI/Jupyter Plugin to monitor Jobs.

```golang
type TrainerConfig struct {

  // Docker image for the Trainer.
  Image string `json:"image,omitempty"`

  // Command for the training container.
  // Validate that command contains torchrun or mpirun.
  Command []string `json:"command,omitempty"`

  // Args for the training container.
  Args []string `json:"args,omitempty"`

  // Env for the training container.
  Env []corev1.EnvVar `json:"env,omitempty"`

  // Number of training nodes.
  NumNodes *int32 `json:"numNodes,omitempty"`

  // Resource for each node.

>   ResourcesPerNode []corev1.resoruces `json:"resourcesPerNode,omitempty"`

Author: @vsoch Do you see a use-case when resources for MPI launcher should be changed frequently of every Job submission ?

>   ResourcesPerNode []corev1.resoruces `json:"resourcesPerNode,omitempty"`

Reviewer 3: For the launcher, if it's just an extra pod/node that exists to send everything off (and the host isn't included in the set to do work, at least given the design of the MPI operator is like this) it could be relatively small in terms of resources. It probably wouldn't need that much changing between job submissions under these circumstances. If you have more of the Flux Operator design where the launcher is also included in the hostlist and does work, then you'd likely want it to have the same resources as the other nodes.

In terms of general resources for the job between submissions, that will definitely change, but I think you are just asking about the MPI Operator launcher. @alculquicondor can likely comment.

>   ResourcesPerNode []corev1.resoruces `json:"resourcesPerNode,omitempty"`

Author: In that case, we have this:
- When launcher works as Node with index 0, resources are the same for every Job replica.
- When launcher is a coordinator, resources will be pre-defined in the blueprint.

In that case, I don't think we should allow user to set resources for coordinator in the `TrainJob` API.

>   ResourcesPerNode []corev1.resoruces `json:"resourcesPerNode,omitempty"`

Reviewer 3: I agree - and minimally since we understand these two designs, we can omit that specification of resources until there is reason to add it.

   </td>
  </tr>
  <tr>
   <td><code>.image</code>
   </td>
   <td><code>.spec.replicatedJobs[name=’Node’].template.spec.template.spec.containers[name=’trainer’].image</code>
   </td>
  </tr>
  <tr>
   <td><code>.command</code>
   </td>
   <td><code>.spec.replicatedJobs[name=’Node’].template.spec.template.spec.containers[name=’trainer’].command</code>
   </td>
  </tr>
  <tr>
   <td><code>.args</code>
   </td>
   <td><code>.spec.replicatedJobs[name=’Node’].template.spec.template.spec.containers[name=’trainer’].args</code>
   </td>
  </tr>
  <tr>
   <td><code>.env</code>
   </td>
   <td><code>.spec.replicatedJobs[name=’Node’].template.spec.template.spec.containers[name=’trainer’].env</code>
   </td>
  </tr>
  <tr>
   <td><code>.numNodes</code>
   </td>
   <td><code>.spec.numNodes</code>
   </td>
  </tr>
  <tr>
   <td><code>.resourcesPerNode</code>
   </td>
   <td><code>.spec.replicatedJobs[name=’Node’].template.spec.template.spec.containers[name=’trainer’].resources</code>
   </td>
  </tr>
</table>

### The Dataset Config API

The `DatasetConfig` represents the APIs that data scientists can use to configure dataset location.

> ```golang

Author: What parameters from your perspective should be part of `DatasetConfig` other than `storageUri` ?

> ```golang

Reviewer 3: At least when I've used models recently, I've needed to add Kaggle credentials (username / token), but I think that is already accounted for. If the storageURI includes a subpath then that is also covered. If there are different mirrors, that is likely either hidden or also covered in the URI. I can't think of anything else now but will post if I do!

> ```golang

Author: Makes sense, so initially adding just `storageUri` and key-value parameters make sense.

    // (Optional). We can support the Iceberg dataset using PyIceberg.
    IcebergDatasetProvider *kubeflowv1.IcebergDatasetProvider `json:"iceberg,omitempty"`
}

type HFDatasetProvider struct {
  // Path to the HF dataset. For example: yelp-review-full
  RepoId string `json:"repoId"`

  // Whether the dataset needs to be splitted: train/val. E.g. split=train[:50000]
  Split string `json:"split,omitempty"`

  // Secret must contain HF_TOKEN secret.
  // If the secret object contains more than one secret, all secrets are passed.
  AccessTokenSecretRef corev1.SecretReference `json:"accessToken,omitempty"`
}

type S3DatasetProvider struct {
  // S3 endpoint.
  EndpointUrl string `json:"endpointUrl,omitempty"`

  // Name of the S3 bucket.
  BucketName string `json:bucketName`

   // Path to the dataset. All files will be downloaded in that path.
  Path string `json:path`

  // Secret must contain AWS_ACCESS_KEY_ID and AWS_SECRET_ACCESS_KEY secret.
  // If the secret object contains more than one secret, all secrets are passed.
  // Otherwise, IRSA can be used to access S3.
  AccessTokenSecretRef corev1.SecretReference `json:"accessToken,omitempty"`
}
```

> The following table explains how `TrainingRuntime` parameters will be overridden with the

Author: Yeah, it should be tables, since we have 2 providers.

```yaml
.spec.replicatedJobs[name='Initializer'].template.spec.template.spec.containers[name=dataset-initializer]
```

#### The HuggingFace Provider

For the HuggingFace provider environment variable `PROVIDER=hf`.

<table>
  <tr>
   <td><strong><code>DatasetConfig</code> Parameter</strong>
   </td>
   <td><strong><code>TrainingRuntime</code> Parameter</strong>
   </td>
  </tr>
  <tr>
   <td><code>.huggingface.repoId</code>
   </td>
   <td><code>.env[REPO_ID]</code>
   </td>
  </tr>
  <tr>
   <td><code>.huggingface.split</code>
   </td>
   <td><code>.env[SPLIT]</code>
   </td>
  </tr>
</table>

#### The S3 Provider

For the S3 provider environment variable `PROVIDER=s3`.

<table>
  <tr>
   <td><strong><code>DatasetConfig</code> Parameter</strong>
   </td>
   <td><strong><code>TrainingRuntime</code> Parameter</strong>
   </td>
  </tr>
  <tr>
   <td><code>.s3.endpointUrl</code>
   </td>
   <td><code>.env[ENDPOINT_URL]</code>
   </td>
  </tr>
  <tr>
   <td><code>.s3.bucketName</code>
   </td>
   <td><code>.env[BUCKET_NAME]</code>
   </td>
  </tr>
  <tr>
   <td><code>.s3.path</code>
   </td>
   <td><code>.env[PATH]</code>
   </td>
  </tr>
</table>

### The Model Config API

> The `ModelConfig` represents the APIs that data scientists can use to configure pre-trained model

Author: >This gets tricky very fast because not only are we now considering custom user preferences (e.g., "save my output to s3 under this path with these credentials" 

I think, you are right, since user also might want to configure some parameters to export model. 
We might want to design this after initial implementation.

In that case, maybe we should add this to the `ModelConfig`:
```yaml
modelConfig:
  preTrainedModel:
    storageUri: hf://bert-base-uncased
    env:
      - name: TRANSFORMER_TYPE
        value: AutoModelForCausalLM
  trainedModel:
    storageUri: oci://models/trained-bert:v1
    env:
      - name: REGISTRY
        value: ghcr.io
```

However, with `startupPolicyOrder: InOrder` policy we can easily create the appropriate JobSet template for it.

```
apiVersion: kubeflow.org/v2alpha1
kind: ClusterTrainingRuntime
metadata:
  name: torch-tune-llama-7b
spec:
  numNodes: 1
  startupPolicy:
    startupPolicyOrder: InOrder
  replicatedJobs:
    - name: Initializer
      template:
        spec:
          template:
            spec:
              containers:
                - name: dataset-initializer
                  image: docker.io/kubeflow/dataset-initializer
                  env:
                    - name: DATASET_PROVIDER
                      value: hf
                    - name: REPO_ID
                      value: tatsu-lab/alpaca
                  volumeMounts:
                    - mountPath: /workspace/dataset
                      name: dataset-initializer
                - name: model-initializer
                  image: docker.io/kubeflow/model-initializer
                  env:
                    - name: MODEL_PROVIDER
                      value: hf
                    - name: REPO_ID
                      value: meta-llama/Llama-2-7b
                    - name: TRANSFORMER_TYPE
                      value: AutoModelForCausalLM
                  volumeMounts:
                    - mountPath: /workspace/model
                      name: model-initializer
              volumes:
                - name: dataset-initializer
                  persistentVolumeClaim:
                    claimName: dataset-initializer
                - name: model-initializer
                  persistentVolumeClaim:
                    claimName: model-initializer
    - name: Node
      template:
        spec:
          template:
            spec:
              containers:
                - name: trainer
                  image: docker.io/kubeflow/llm-trainer
                  env:
                    - name: MASTER_ADDR
                      value: "pytorch-node-0-0.pytorch"
                    - name: MASTER_PORT
                      value: 29400
                    - name: TRANSFORMER_TYPE
                      value: AutoModelForCausalLM
                    - name: LORA_CONFIG
                      value: |
                        {"peft_type": "LORA", "r": 8, "lora_alpha": 16}
                  command:
                    - torchrun hf_llm_training.py
                  resources:
                    limits:
                      nvidia.com/gpu: 2
                  volumeMounts:
                    - mountPath: /workspace/dataset
                      name: dataset-initializer
                    - mountPath: /workspace/pre-trained-model
                      name: model-initializer
                    - mountPath: /workspace/adapters
                      name: model-exporter
              volumes:
                - name: dataset-initializer
                  persistentVolumeClaim:
                    claimName: dataset-initializer
                - name: model-initializer
                  persistentVolumeClaim:
                    claimName: model-initializer
                - name: model-exporter
                  persistentVolumeClaim:
                    claimName: model-exporter
    - name: Exporter
      template:
        spec:
          template:
            spec:
              containers:
                - name: model-exporter
                  image: docker.io/kubeflow/model-exporter
                  volumeMounts:
                    - mountPath: /workspace/adapters
                      name: model-exporter
              volumes:
                - name: model-exporter
                  persistentVolumeClaim:
                    claimName: model-exporter
```

> The `ModelConfig` represents the APIs that data scientists can use to configure pre-trained model

Reviewer 3: Dumb question - instead of this:

```yaml
modelConfig:
  preTrainedModel:
    storageUri: hf://bert-base-uncased
    env:
      - name: TRANSFORMER_TYPE
        value: AutoModelForCausalLM
  trainedModel:
    storageUri: oci://models/trained-bert:v1
    env:
      - name: REGISTRY
        value: ghcr.io
```
Why would you not have:

```yaml
modelConfig:
  preTrainedModel:
    storageUri: hf://bert-base-uncased
    env:
      - name: TRANSFORMER_TYPE
        value: AutoModelForCausalLM
  trainedModel:
    storageUri: oci://ghcr.io/models/trained-bert:v1
```

Or it could be `storageUri: oras://ghcr.io/models/trained-bert:v1` to indicate pushing to an OCI registry with ORAS. I'm wondering why the registry is a separate variable when it is part of the URI?

> The `ModelConfig` represents the APIs that data scientists can use to configure pre-trained model

Author: Yeah, you are right!
Similar to this KEP: https://github.com/kubernetes/enhancements/pull/4642, we should set this as follows:
```
storageUri: oci://ghcr.io/models/trained-bert:v1
```

> The `ModelConfig` represents the APIs that data scientists can use to configure pre-trained model

Reviewer 8: Since this API is meant to hide the complexity from data scientists, the keys that can be overridden/customized could be exposed for simplicity. For instance, on TrainJob

Instead of
```
modelConfig:
  preTrainedModel:
    storageUri: hf://bert-base-uncased
    env:
      - name: TRANSFORMER_TYPE
        value: AutoModelForCausalLM
  trainedModel:
    storageUri: oci://ghcr.io/models/trained-bert:v1
 ```
 
 could we have
 ```
modelConfig:
  preTrainedModel:
    storageUri: hf://bert-base-uncased
    transformerType: AutoModelForCausalLM   <--- Key that is overridden by user
  trainedModel:
    registry: quay.io    <---- This would change the storageUri to be oci://quay.io/models/trained-bert:v1
 ```
 
 Tracking down which environment variable needs to be modified can be hard. A subset of customizable keys with well-explained documentation could make the API more user-friendly.

> The `ModelConfig` represents the APIs that data scientists can use to configure pre-trained model

Reviewer 3: I like that overall suggestion - but I think I would still do:

```yaml
modelConfig:
 preTrainedModel:
   storageUri: hf://bert-base-uncased
   transformerType: AutoModelForCausalLM
 trainedModel:
   storageUri: oci://quay.io/models/trained-bert:v1
```

> The `ModelConfig` represents the APIs that data scientists can use to configure pre-trained model

Reviewer 6: In the idea of having environment variables `TRANSFORMER_TYPE` like attributes `transformerType`?
If yes, I liked the attributes idea but having a way to use environment variables gives us flexibility.
Maybe we should support both having a defined priority? Example: If we have `transformerType` and `TRANSFORMER_TYPE` the last defined in the YAML has the priority. This is like Linux, the last defined "env" will be used.

Another thing, may we support ConfigMap? 

I know this is a lot for now. Maybe for v2.0.0 we should use `transformerType` and start a new discussion in v2.x.x about having env vars, config maps, etc

> The `ModelConfig` represents the APIs that data scientists can use to configure pre-trained model

Reviewer 8: > I like that overall suggestion - but I think I would still do:
> 
> ```yaml
> modelConfig:
>  preTrainedModel:
>    storageUri: hf://bert-base-uncased
>    transformerType: AutoModelForCausalLM
>  trainedModel:
>    storageUri: oci://quay.io/models/trained-bert:v1
> ```

Yes. If the storageUri had been baked into the blueprint, overriding the registry field in the TrainJob would put in the new storageUri in the final JobSet. Alternatively, the user can provide the entire storageUri to be replaced.

> The `ModelConfig` represents the APIs that data scientists can use to configure pre-trained model

Reviewer 8: > Is the idea of having environment variables `TRANSFORMER_TYPE` like attributes `transformerType`? If yes, I liked the attributes idea but having a way to use environment variables gives us flexibility. Maybe we should support both having a defined priority? Example: If we have `transformerType` and `TRANSFORMER_TYPE` the last defined in the YAML has the priority. This is like Linux, the last defined "env" will be used.
> 
> Another thing, may we support ConfigMap?
> 
> I know this is a lot for now. Maybe for v2.0.0 we should use `transformerType` and start a new discussion in v2.x.x about having env vars, config maps, etc

Agree with the point on flexibility. Though, having to add an environment variable would mean data scientists understand the internal implementation, for ex: TRANSFORMER_TYPE is used as an environment variable and not a command line flag. The most useful config map keys could be similarly converted to attributes. Also, if there are a bunch of env variables (env or config map) that always get added in every trainJob, then it may make sense to create a blueprint with it, as per the overall intention of this proposal.

> The `ModelConfig` represents the APIs that data scientists can use to configure pre-trained model

Author: The main problem with API attributes is that different model providers and dataset providers might have different configurations user can override. For example, S3 provider might have endpoint URL, but HF doesn't not have it.
Thus, this approach is making override more extensible to support various storage initializers:

```yaml
modelConfig:
  preTrainedModel:
    storageUri: hf://bert-base-uncased
    env:
      - name: TRANSFORMER_TYPE
        value: AutoModelForCausalLM
  trainedModel:
    storageUri: oci://quay.io/models/trained-bert:v1
datasetConfig:
  storageUri: s3://datasets/yelp-review
  env:
    - name: ENDPOINT_URL
      value: s3.custom.com
```
Eventually, all of those env parameters should exist in the blueprint `model-initializer` and `storage-initializer` containers under `Initializer` Job.

@tenzen-y @shravan-achar @vsoch @diegolovison @johnugeorge @deepanker13 Any thoughts on how we can construct those API?

> The `ModelConfig` represents the APIs that data scientists can use to configure pre-trained model

Reviewer 3: If I understand the example from [this comment](https://github.com/kubeflow/training-operator/pull/2171#discussion_r1681926971), it seems like the `env` section is being used as a place to put random variables hat are relevant for the different Initializer jobs, and one example subset is related to the storageUri. I don't think I like (or see the need for) having redundancy there - more specifically any variable that can be represented in a Uri should be there, and it's easier to have one source of truth (vs. allowing an environment variable that goes into the Uri, and then the user provides both and it's not clear what the final result will be). In the case of variables for the Initializer that have no overlap (no other place to put them) then we are just using `env` as a storage place for key value pairs. In terms of what it looks like for the user, if we wanted something a bit more straight forward, why not just allow key value pairs, always string / string, that can be customized? Even storageURI could be one of them, e.g.,:

```yaml
modelConfig:
  preTrainedModel:
    storageUri: hf://bert-base-uncased
    transformerType: AutoModelForCausalLM
  trainedModel:
    storageUri: oci://quay.io/models/trained-bert:v1
datasetConfig:
  storageUri: s3://datasets/yelp-review
  endpointUrl: s3.custom.com
```
Or if storageURI is special (or not a variable in the key/value set) make them explicitly separate:

```yaml
modelConfig:
  preTrainedModel:
    storageUri: hf://bert-base-uncased
    custom:
      transformerType: AutoModelForCausalLM
  trainedModel:
    storageUri: oci://quay.io/models/trained-bert:v1
datasetConfig:
  storageUri: s3://datasets/yelp-review
  custom:
    endpointUrl: s3.custom.com
```
And then the rule would be that anything under custom has a straight forward mapping to an environment variable, such as replacing uppercase or dashes with an underscore, and then making the entire thing uppercase. Then the underlying container would know how to use those. For the actual environment mapping, I think we want the final variable names to be namespaced. E.g.,:

```bash
DATASET_CONFIG_ENDPOINT_URI=s3.custom.com
MODEL_CONFIG_TRANSFORMER_TYPE=AutoModelForCasualLM
```
That way, I can do `env | grep PREFIX` to find an entire set for a particular type, and if necessary, they can exist in the same space together (where there could be overlapping names).

I can think more about the yaml design - I think I want to better understand cases (outside of storage URIs) these variables are used and needed.

> The `ModelConfig` represents the APIs that data scientists can use to configure pre-trained model

Author: Yes, key-value pairs also should work. I know that KServe is doing this in the `storageSpec` API:
https://github.com/kserve/kserve/blob/master/pkg/apis/serving/v1beta1/predictor.go#L99-L101
Let me try to refine the API and we can discuss it again.

> The `ModelConfig` represents the APIs that data scientists can use to configure pre-trained model

Reviewer: This KEP has quite a lot in it. My gut is thinking that ModeConfig and DataSetConfig could be its own API that lives in Kubeflow/Common or somewhere else. 

I'd love to see the JobSet + TrainJob interface merged eventually and I worry that there will be a lot of back and forth on ModeConfig/DatasetConfig and that may block the progression of this feature.

I don't know where I found it but someone has already asked about this for pipelines and I'm sure notebooks would also want to use the same API. Maybe we consider splitting this from this for now?

> The `ModelConfig` represents the APIs that data scientists can use to configure pre-trained model

Author: Yes, @diegolovison and @tarilabs discuss it here: https://github.com/kubeflow/training-operator/pull/2171#discussion_r1681096006.
We can always improve these API in the future version, that is why we call them `v2alpha1` initially.

I just think that the `datasetConfig` and `modelConfig` API will be very beneficial for us to work on, since majority of users are doing fine-tuning today.
Especially, this will be useful when user runs multi-note multi-worker training, when data needs to be distributed across multiple Pods.

About other Kubeflow components, I know that KServe is already using storage initializer to download model.

I don't think we will get any blockers from other Kubeflow WGs to implement it.

On the planning side, this APIs should be implemented after initial support of simple PyTorch Distributed Runtime.

> The `ModelConfig` represents the APIs that data scientists can use to configure pre-trained model

Reviewer: Sounds good to me. I tend to making KEPs/designs as simple as possible so we don't get bog down on details.

input and output location.

```golang
type ModelConfig struct {
    // One of the following can be set.
    HFModelProvider *kubeflowv1.HFModelProvider `json:"huggingface,omitempty"`

    // Potential output location for fine-tuned/trained model.
    OutputArtifact string
}

type HFModelProvider struct {
    // Path to the pre-trained model. google-bert/bert-base-uncased

>     RepoID string `json:"repoID"`

Author: I think, we just try to keep this API consistent to the HuggingFace hub `load`: https://huggingface.co/docs/huggingface_hub/en/package_reference/cards#huggingface_hub.RepoCard.load.
On the other side, `from_pretrained` call it model Id: https://github.com/huggingface/transformers/blob/v4.42.0/src/transformers/modeling_utils.py#L2850.

However, if we agree to use `storageUri`, we don't need these differences.

          node
      containers:
        - name: user-identity
          value: 123
        - name: trainer
          volumeMounts:
            - name: user-123-volume
              mountPath: /workspace
      volumes:
        - name: user-123-volume
          persistentVolumeClaim:
            claimName: user-123-volume
```

## The Training Runtime API

The `TrainingRuntime` is the pre-created configurations of model training on the cluster,
representing as blueprints. For example, Elastic PyTorch training, MPI DeepSpeed configuration,
BERT LLM Fine-Tuning.

These blueprints can be deployed within the Training Operator control plane and stored in a Kubeflow
public repository that users can apply to their clusters.

Platform or ML engineers can tweak existing blueprint, based on their requirements. For example,
using custom configurations.

The Kubeflow Training Operator can maintain more Training Runtimes when the community is ready to
support them. For example, runtimes for [Jax](https://jax.readthedocs.io/en/latest/index.html) or
[MLX](https://ml-explore.github.io/mlx/build/html/index.html). Initially, we will support: PyTorch,
MPI, TensorFlow, XGBoost, and PaddlePaddle.

> The `TrainingRuntime` is immutable, and so to make a change, a new version of the `TrainingRuntime`

Author: Sure, let me try to think about another diagram to explain it better.

> The `TrainingRuntime` is immutable, and so to make a change, a new version of the `TrainingRuntime`

Author: @vsoch What do you think about this new diagram ?

> The `TrainingRuntime` is immutable, and so to make a change, a new version of the `TrainingRuntime`

Reviewer 3: It's great! I can give some nits, but overall it hugely helps to show the roles and interactions. My nits:

- Should "Fetch Spec" be a sort of bi-directional thing? We fetch (and retrieve)
- The training nodes (job) includes the headless service and pods? They are displayed as separate. From a user perspective, once you make the JobSet, it gets deployed as a unit that contains all of those components. Maybe the diagram could reflect that, versus now it looks like Training Nodes get changed into a headless service.

must be created and then changing the `TranJob` to point to the new version.
This provides control as to how changes to runtimes propagate to existing training jobs.
For example, when training is running for a long time (e.g. 1-2 months).

In the future implementation, we will introduce a revision control mechanism similar to
[Kubernetes Deployment](https://kubernetes.io/docs/concepts/workloads/controllers/deployment/#updating-a-deployment)
to control versions of `TrainingRuntime` and enable rolling updates.

> We are going to create two CRDs: `TrainingRuntime` and `ClusterTrainingRuntime`. These runtimes have

Author: Let's continue in this 🧵 https://github.com/kubeflow/training-operator/pull/2171#discussion_r1680412433

exactly the same APIs, but the first one is the namespace-scoped, the second is the cluster-scoped.
If `trainingRuntimeRef` in `TrainJob` has the namespace, controller will use the `TrainingRuntime`,
otherwise it will use the `ClusterTrainingRuntime`.

```golang
type TrainingRuntime struct {
    metav1.TypeMeta   `json:",inline"`
    metav1.ObjectMeta `json:"metadata,omitempty"`

    // Framework specific parameters.
    MLSpec *MLSpec `json:"mlSpec,omitempty"`

    // Number of nodes to execute training.
    NumNodes int `json:"numNodes,omitempty"`

    // JobSet spec.
    JobSetSpec *batchv1.JobSetSpec `json:",inline"`

    // For gang-scheduling using volcano or scheduler plugins, supported for all frameworks.

>     GangScheduler *kubeflowv1.GangScheduler `json:"gangScheduler,omitempty"`

Reviewer 3: Just a curiosity, for the ML user, are they more familiar with Gang scheduling? I know gang is common in the developer community, but how far does that extend? Why not call it GroupScheduling?

Author: Usually, MLOps engineers should be aware of these terms, but I let @tenzen-y to speak more on this.

Author: @tenzen-y @vsoch Should we just call it `scheduler` to make it more generic?

Reviewer 5: I think that `gangScheduler` should be fine since this is not related to scheduler field to PodSpec.
This is only for PodGroup reconciliations. The `scheduler` field name could bring users some confusions.

Author: I am wondering, do we want to re-use this parameter not only for gang-scheduling ?

For example, if users want to use custom scheduler (YuniKorn or Volcano) to schedule TrainJobs with 1 node ?
For example, Spark Operator introduces YuniKorn support recently: https://github.com/kubeflow/spark-operator/pull/2107 (cc @yuchaoran2011 @vara-bonthu @shravan-achar @jacobsalway @ChenYi015)

Could it be useful for ML Training Jobs as well ?

Author: We can consider using this API to change pod scheduler: https://github.com/kubernetes/api/blob/master/core/v1/types.go#L3874

}

// One of the specs can be selected.
type MLSpec struct {

    // Custom Spec for Torch
    TorchSpec *TorchSpec `json:"torchSpec,omitempty"`

    // Custom Spec for MPI
    MPISpec *MPISpec `json:"mpiSpec,omitempty"`
}
```

### The Gang Scheduler API

Gang scheduler plugin is used to create appropriate `PodGroup` for Volcano or scheduler plugins.

```golang
type GangScheduler struct {
    // Plugin for gang scheduling.
    Plugin *GangSchedulerPlugin `json:plugin,omitempty"`

    // Time threshold to schedule PodGroup for gang scheduling.
    ScheduleTimeoutSeconds string `json:scheduleTimeoutSeconds,omitempty"`
}

type GangSchedulerPlugin string

const (
    GangSchedulerPluginVolcano GangSchedulerPlugin = "volcano"

>     GangSchedulerPluginSP      GangSchedulerPlugin = "scheduler-plugins"

Reviewer 3: SP == service provider?

Author: It is when Scheduler Plugins are used for Gang Scheduling: https://github.com/kubernetes-sigs/scheduler-plugins?tab=readme-ov-file
Any thoughts on how we can make this name more clear ?

Reviewer 3: What does it explicitly read to? 

GangSchedulerPluginSP == GangSchedulerPlugin Scheduler Plugin?

Author: Yes, it is for Scheduler Plugins.

Reviewer 3: Perhaps just remove “SP” then, because it’s redundant - or add “Name” to one of the variables if needed.

we won't support them in `TorchSpec`. We can introduce them in the future if users will require them.

```golang
// TorchSpec represents the configuration for PyTorch.
type TorchSpec struct {

    // Number of Procs per Node.
    NumProcPerNode int `json:"numProcPerNode,omitempty"`

    // Used for single-node multi-worker training
    Standalone bool `json:"standalone,omitempty"`

    // Torch Elastic Policy.
    ElasticPolicy *TorchElasticPolicy `json:"elasticPolicy,omitempty"`
}

// If the Elastic Policy is set, the numNodes parameter is ignored.
// --nnodes=minNodes:maxNodes
type TorchElasticPolicy struct {

    // The limits to restart TrainJob.
    // Insert it to the JobSet.spec.failurePolicy.maxRestarts
    MaxRestarts *in32 `json:"maxRestarts,omitempty"`

    // Min number of nodes for HPA and torchrun.
    MinNodes *in32 `json:"minNodes,omitempty"`

    // Max number of nodes for HPA and torchrun.
    MaxNodes *in32 `json:"maxNodes,omitempty"`

    // Metrics for scale up and down replicas.
    Metrics []autoscalingv2.MetricSpec `json:"metrics,omitempty"`
}
```

### The MPI Spec API

The `MPISpec` API represents the configuration for training using MPI orchestration.

> E.g. creation of host-files and SSH keys. Using MPI might be more efficient for training on HPC

Reviewer 3: HPC clusters will usually have a workload manager that handles the MPI bootstrap and gets rid of the need for a hostfile and ssh. For example, flux uses zeromq for communication and already is aware of its hosts. One thing I want to keep in mind is, because we are anticipating running user space Kubernetes on our systems for the future, we want these abstractions to map nicely. For that I would say to leave this open and flexible to support not just different MPI interface commands, but also entirely different workload managers. As a side note, I am very happy to see MPI here :)

Author: Sure, I think, we can support other workload managers in our TrainingRuntime in the future (e.g. Flux, Slurm).
For example, I know that right now running PMIx is not possible with MPI Operator: https://github.com/kubeflow/mpi-operator/issues/12.

@vsoch Any thoughts on how we can design our `MLSpec` to make it more generic for other workload managers ?

Reviewer 3: To step back - what is distinct about an MPISpec? Right now it is the launcher, and need to setup ssh? If we were to deploy a workload manager, and one that didn't require a special launcher node, the resources could be equivalent, and the main difference would be that one node (the leader) is running a slightly different command than the workers (followers). I think what we should do is write down explicitly the features that are unique to an MPISpec, figure out the generic design, and try to remove needing to name specific implementations to the largest degree possible. Arguably, an MPISpec is valid for "something that uses MPI" but should be flexible to change in that landscape over time (and even supporting the workload managers).

Author: The main motivation behind `MLSpec` is to simplify experience for Platform Engineers on how to configure various parameters for different runtimes. For example, `ElasticPolicy` parameter is only applicable for PyTorch since TensorFlow distributed doesn't support it.
For the workload managers, I am wondering do they (Slurm, Flux, and others) are always use MPI to run distributed job, or we can run them without MPI. For example, Slurm + `torchrun`. cc @kuizhiqing

Reviewer 3: Yes, of course you can run non MPI jobs under a workload manager! I didn't use MPI for my entire graduate school career (it was more machine learning and biosciences pipelines). You can use the same features (the ability to launch and schedule with batch, or to sub-instances in a graph) in absence of MPI. As as example, I've run torchrun with the Flux Operator ([example](https://github.com/flux-framework/flux-operator/tree/6a223538519adafd2785ebe4cd92afc6d34d6444/examples/machine-learning/pytorch)). 

Workload managers are not explicitly wed to MPI, but because MPI is a bread and butter HPC communication interface ([detail](https://www.osti.gov/servlets/purl/1860774), it's standard that the workload managers support running these apps optimally (low latency, bootstrap, etc). As a side note on elasticity, we have support for a faux kind of elasticity for the Flux Operator (adding nodes dynamically) and should have full support in bare metal flux (without Kubernetes) to grow and shrink jobs soon.

Author: That makes sense, so in that case maybe in the future we can introduce another union type in our MLSpec to support it (e.g. FluxSpec, SlurmSpec).
That gives Platform Engineers ability to create training blueprints utilising various workload managers capabilities.

Reviewer 3: Absolutely. I can offer to help with that, when the time comes.

clusters or for some ML frameworks (e.g. [MLX distributed with MPI](https://ml-explore.github.io/mlx/build/html/usage/distributed.html)).

We will fully migrate to the MPI Operator V2 functionality as part of this KEP.
Check [the proposal for the MPI V2 APIs.](https://github.com/kubeflow/mpi-operator/blob/master/proposals/scalable-robust-operator.md)

```golang
type MPISpec struct {
    // Number of Procs per Node.
    NumProcPerNode int `json:"numProcPerNode,omitempty"`

    // MPI Implementation to create appropriate host-files.
    // Can be one of OpenMPI, Intel, or MPICH.
    MPIImplementation MPIImplementation `json:"mpiImplementation,omitempty"`

    // Directory where SSH keys are mounted.
    SSHAuthMountPath string `json:"SSHAuthMountPath,omitempty"`
}

type MPIImplementation string

> const (

Reviewer 3: To continue my comment above, if we use a workload manager the MPI implementation may not be needed at this detail. The application would be built with a variant, definitely, but bootstrapped by the workload manager.

    MPIImplementationOpenMPI MPIImplementation = "OpenMPI"
    MPIImplementationIntel   MPIImplementation = "Intel"
    MPIImplementationMPICH   MPIImplementation = "MPICH"
)
```

### Supported Runtimes by Community

Kubeflow community are planning to support the following runtimes.

#### PyTorch Distributed Runtime

Initially, we will maintain only multi-node multi-worker runtime and PyTorch Elastic.

```yaml
apiVersion: kubeflow.org/v2alpha1

> kind: ClusterTrainingRuntime

Reviewer 3: Yeah thinking this through more - let's talk about why different abstractions are needed for cluster vs. not. I think this design would be much simpler (easier for the user to use) if there were just one.

Reviewer 8: Agreed with the idea that having two layers is complex.

Reading through the doc, I think ClusterTrainingRuntime is meant to be something set by Kubeflow community or designed by platform engineers in an org incorporating all the best practices. This is cluster scoped. This may not be runnable as is and needs to be further tuned or scoped down for team specific settings like model provider or custom init containers. 

ML teams may just have access to namespace-scoped resources so they end up copying the blueprint, fill in more defaults and make them available with other MLEs in their team to try with different trainerConfigs. Since these blueprints are specific to a particular team of engineers, it may not fit the cluster scope. This part needs to be optional however ie. TrainerConfigs can reference a cluster-scoped ClusterTrainingRuntime.

The idea does have merits (similar to ClusterRole and Role concepts) but is certainly increases complexity.

Author: @tenzen-y Can share concrete on-prem use-case why do we need `TrainingRuntime` and `ClusterTrainingRuntime`.

TLDR: If you want to share blueprint between all Kubernetes namespaces, you might want to use `ClusterTrainingRuntime`.

I think, that gives flexibility for Platform Engineers to configure blueprints in the Kubernetes cluster.
For example, if you want to create unify MPI runtime per cluster, you should use `ClusterTrainingRuntime`.
On the other side, if you have different namespaces in your cluster (e.g. `ad-hoc` and `batch`), you might consider to use different runtimes for every namespace. For example, using different accelerators or node selectors.

I understand your confusion in this comment: https://github.com/kubeflow/training-operator/pull/2171#discussion_r1680385287, but let's continue discussion in this 🧵 to see the value of both runtimes.

A few reference:
- KServe `ServingRuntime` and `ClusterServingRuntime`: https://kserve.github.io/website/latest/modelserving/servingruntimes/
- Cert Manager `Issuer` and `ClusterIssuer`: https://cert-manager.io/docs/concepts/issuer/#namespaces

@ahg-g would love to get your feedback as well.

Reviewer 5: Some use cases here: 
- `ClusterTrainingRuntime` are defined by a cluster admin or platform developers and then researchers specify the clusterTrainingRuntime name in the TrainJob API.
- `TrainingRuntime` are defined by researchers within the their namespaces (tenants).

If we do not provide the `ClusterTrainingRuntime` API, admins need to create the TrainingRuntime API in all user namespaces. It is terrible since the multitenant cluster has hundred and thousand namespases.

metadata:
  name: torch-distributed-multi-node
spec:
  mlSpec:
    torch:
      numProcPerNode: 5
  replicatedJobs:
    - name: node
      template:
        spec:
          template:
            spec:
              containers:
                - name: trainer
                  image: docker.io/kubeflow/pytorch-mnist
                  env:
                    - name: MASTER_ADDR
                      value: "pytorch-node-0-0.pytorch"
                    - name: MASTER_PORT
                      value: 29400
                  command:
                    - torchrun train.py
```

Example of usage:

```yaml
apiVersion: kubeflow.org/v2alpha1
kind: TrainJob
metadata:
  name: torch-distributed-multi-node
  namespace: tenant-alpha
spec:
  trainingRuntimeRef:

>     name: pytorch-distributed

Reviewer 3: Should this match the pytorch name above? Or is that not how these are related?

>   trainerConfig:

Reviewer 3: This would be passed to the scheduler (eventually) - I'm a bit tired so the path there isn't totally clear - I'm guessing it will eventually be turned into units of pods with resource requests.

    resourcesPerNode:
      requests:
        nvidia.com/gpu: 1
    args:
      - num-epochs=5
```

#### PyTorch Elastic Runtime

Training runtime for PyTorch Elastic:

```yaml
apiVersion: kubeflow.org/v2alpha1
kind: ClusterTrainingRuntime
metadata:
  name: torch-distributed-elastic
spec:
  mlSpec:
    torchSpec:

>       elasticPolicy:

Reviewer 3: Oh I missed this from earlier - elasticPolicy is for the autoscaler? Is this the built in / expected metrics like CPU or does this also support custom metrics / Prometheus?

Author: We use this to create HPA based similar as here: https://github.com/kubeflow/training-operator/blob/master/pkg/controller.v1/pytorch/hpa.go#L80. As long as HPA can be triggered using Prometheus metrics, we can support it.

PyTorch [supports dynamic changes of Node during model training](https://pytorch.org/docs/stable/elastic/run.html#elastic-min-1-max-4-tolerates-up-to-3-membership-changes-or-failures), we want to implement this functionality.

I know there is pending issue to support Elastic JobSet: https://github.com/kubernetes-sigs/jobset/issues/463.

cc @tenzen-y @kannon92 @ahg-g

Reviewer 3: Gotcha, and just to be clear, elastic here means that we are actually taking a running job and extending the resources to it. We aren't just changing the size of resources to accommodate more jobs.

        minNodes: 5
        maxNodes: 10
        metrics:
          - type: Resource
            resource:
              name: cpu
              target:
                type: Utilization
                averageUtilization: 80
  replicatedJobs:
    - name: node
      template:
        spec:
          template:
            spec:
              containers:
                - name: trainer
                  image: docker.io/kubeflow/pytorch-mnist
                  env:
                    - name: MASTER_ADDR
                      value: "pytorch-node-0-0.pytorch"
                    - name: MASTER_PORT
                      value: 29400
                  command:
                    - torchrun train.py
```

#### Additional PyTorch Runtimes

> The following runtimes can be maintained in the future.

Reviewer 3: Should this be part of this initial KEP if it's TBA / for the future/ maybe very likely to change?

                    - name: LORA_CONFIG
                      value: |
                        {"peft_type": "LORA", "r": 8, "lora_alpha": 16}
                  command:
                    - torchrun hf_llm_training.py
                  resources:
                    limits:
                      nvidia.com/gpu: 2
                  volumeMounts:
                    - mountPath: /workspace/dataset
                      name: dataset-initializer
                    - mountPath: /workspace/model
                      name: model-initializer
              volumes:
                - name: dataset-initializer
                  persistentVolumeClaim:
                    claimName: dataset-initializer
                - name: model-initializer
                  persistentVolumeClaim:
                    claimName: model-initializer
```

#### MPI Runtime

For MPI, we can add support the `DeepSpeed` runtimes.

Example of simple OpenMPI runtime:

```yaml
apiVersion: kubeflow.org/v2alpha1
kind: ClusterTrainingRuntime
metadata:
  name: mpi-simple
spec:
  mlSpec:
    mpi:
      mpiImplementation: OpenMPI
      numProcPerNode: 5
  numNodes: 5
  replicatedJobs:

>     - name: Launcher

Reviewer 3: I'm a little worried about this design because MPI doesn't always need a node that is explicitly for a launcher - that's a design that the MPI operator has taken (and makes sense, it's part of how MPI design is described), but for example, the lead broker of the flux operator does the launch with that node handling the same work as a worker.

Reviewer 3: Follow up question - could the Flux Operator be handled here? High level, it's an MPI launcher (but actually is an entire HPC cluster that simply can be used to run one MPI job - it's an indexed job + headless service under the hood, with flux orchestration and nodes connected via a tree based overlay network with zeromq).

Author: >I'm a little worried about this design because MPI doesn't always need a node that is explicitly for a launchel

Yes, I know we have an API to run launcher as worker in MPI-Operator: https://github.com/kubeflow/mpi-operator/blob/master/pkg/apis/kubeflow/v2beta1/types.go#L161.

I want to hear @tenzen-y and @alculquicondor opinion on this. We didn't get a chance to deeply discuss how MPI Runtime should look like.

Author: >Follow up question - could the Flux Operator be handled here? High level, it's an MPI launcher (but actually is an entire HPC cluster that simply can be used to run one MPI job - it's an indexed job + headless service under the hood, with flux orchestration and nodes connected via a tree based overlay network with zeromq)

Yes, I think that is possible. Do you know if we can construct the Flux cluster for batch using JobSet spec or we need to use Flux operator mini-cluster spec ?
```
 JobSetSpec *batchv1.JobSetSpec `json:",inline"`
 ```

Reviewer 3: The Flux Operator (under the hood) is an indexed job, so yes I could very easily turn that into a JobSet. I actually did a design with JobSet early on, but didn't wind up converting entirely for simple reasons - JobSet wasn't part of Kubernetes core (and would require the user to install an additional thing), and JobSet makes the DNS names much longer, and (in my testing) it often led to errors depending on the name of the job that I chose. Since I just needed the IndexedJob and then the headless service, it was more straight forward to just implement that.

Please put me down to implement Flux Framework into this setup - the nested hierarchical scheduler can handle a lot of really cool stuff within a single job, and the zeromq bootstrap is an improvement over ssh.

Author: Would be interesting to see the benchmarks to compare zmq over ssh for distributed model training. cc @tenzen-y

Reviewer 3: I can definitely help with that, if there is interest to run experiments. We compared an HPC proxy app (LAMMPS) to the MPI operator, now that was a long time ago (early 2023) and there was about a 5% difference in means, and you can imagine that would compound with runtime (our runs were very short). That said, both have changed immensely since then!

I have a hard time getting the MPI Operator working, but if we wanted to decide on a workflow, I could offer to set it up and run in the Flux Operator. We would need to choose a cloud that has a suitable network and resources for the work. If it's ML then probably Google would work - I have credits currently.

Reviewer 5: I still think that both MPIJob modes (runLauncherAsWorker and specific launcher) would be helpful for the GPU and CPU mixed clusters. Additionally, the launcher allows us to easily migrate from MPIJob v2 to TrainJob.

easily.

Please also see the [Kubernetes JobSet and Kubeflow Training Operator collaboration document](https://docs.google.com/document/d/1C2ev7yRbnMTlQWbQCfX7BCCHcLIAGW2MP9f7YeKl2Ck/edit?usp=sharing).

### User Value

In addition to the above motivation, we will address the following user feedback while implementation:

- Confusion around Workers: https://github.com/kubeflow/training-operator/issues/1790
- Support batch/v1 `Job` features: https://github.com/kubeflow/training-operator/issues/1718
- ExitCodes for PodFailurePolicy: https://github.com/kubeflow/training-operator/issues/1749
- Migrate to MPI V2 API: https://github.com/kubeflow/training-operator/issues/1906

### Personas

We can identify the following personas of Training Operator:

1. **DevOps Engineer**. They are familiar with Kubernetes concepts and they know how to manage the
   Kubernetes workloads. Usually, they are not experts in ML frameworks and ML algorithms.
1. **MLOps Engineer**. They are familiar with ML frameworks and they know how to configure
   distributed PyTorch settings such as rendezvous backends or MPI configuration. Usually, they are
   not experts in Kubernetes and ML algorithms.

> 1. **Data Scientists**. They create model architectures and advanced ML algorithms to train models.

Author: Make sense.

	// Configuration for training dataset
	DatasetConfig *DatasetConfig `json:"datasetConfig,omitempty"`

	// Configuration for the pre-trained model and location for model output
	ModelConfig *ModelConfig `json:"modelConfig,omitempty"`

	// Custom metadata to apply for Job, JobSet, etc.
	Labels      map[string]string `json:"labels,omitempty"`
	Annotations map[string]string `json:"annotations,omitempty"`
}

type TrainingRuntimeRef struct {
	// Name for the training runtime.
	Name string `json:"name"`
	// Namespace for the runtime.
	// If namespace is set, TrainingRuntime is used. Otherwise, ClusterTrainingRuntime is used.
	Namespace string `json:"namespace,omitempty"`
}

type TrainJobStatus struct {
	// Conditions for the TrainJob
	Conditions []metav1.Condition `json:"conditions,omitempty"`
}

```

This table explain rationale for each `TrainJob` parameter:

<table>
  <tr>
   <td><strong>Parameter</strong>
   </td>
   <td><strong>What is it ?</strong>
   </td>
  </tr>
  <tr>
   <td><code>TrainingRuntimeRef</code>
   </td>
   <td>Reference to the existing <code>TrainingRuntime</code> that pre-deployed by platform engineers
   </td>
  </tr>
  <tr>

>    <td><code>Trainer</code>

Reviewer 9: I don't think we defined above what exactly a Trainer is, so that would be useful in my opinion.

Author: @franciscojavierarceo I think, I am trying to explain it here: https://github.com/kubeflow/training-operator/blob/8dd9c85ce81b1ff9e1576e584e698408c56c162b/docs/proposals/2170-kubeflow-training-v2/README.md#the-trainer-api.
Do you think it is not sufficient ?

Reviewer 9: Yeah that works, I suppose I meant that in the ordering of the document it may make sense to move the definition first so that when someone sees the tables they have that knowledge first. That's also a bit of a silly nit, so feel free to disregard. Thanks for putting this together!

   </td>
   <td>Configuration for the Trainer such as image, number of nodes, accelerators.
   </td>
  </tr>
  <tr>
   <td><code>ModelConfig</code>
   </td>

>    <td>Configuration for the pre-trained model and location for model output

Reviewer 9: ```suggestion
   <td>Configuration of the pre-trained model and the location of the model output
```

   </td>
  </tr>
  <tr>
   <td><code>DatasetConfig</code>
   </td>

>    <td>Configuration for the dataset that will be used to train or fine-tune model

Reviewer 9: ```suggestion
   <td>Configuration for the dataset that will be used to train or fine-tune the model
```

   </td>
  </tr>
  <tr>
   <td>Labels and Annotations
   </td>

>    <td>Custom metadata that needs to be applied to the <code>TrainJob</code> resources: JobSet, Job, Pods.

Reviewer 9: ```suggestion
   <td>Custom metadata that needs to be applied to the <code>TrainJob</code> resources: JobSet, Job, and Pods.
```

   </td>
  </tr>
  <tr>
   <td><code>PodSpecOverrides</code>
   </td>
   <td>Custom overrides that are specific to the <code>TrainJob</code> and need to be applied to the
    <code>TrainJob</code> resources. For example, the user identity. Usually, it is managed by

>     custom admission webhooks that inject data to the <code>TrainJob</code> after user creates it

Reviewer 9: ```suggestion
    custom admission webhooks that inject data to the <code>TrainJob</code> after the user creates it
```

>     via Python SDK or <code>kubectl</code>

Reviewer 9: ```suggestion
    via the Python SDK or <code>kubectl</code>
```

   </td>
  </tr>
</table>

### Example of TrainJob

```yaml
apiVersion: kubeflow.org/v2alpha1
kind: TrainJob
metadata:
  name: torch-ddp
  namespace: tenant-alpha
spec:
  trainingRuntimeRef:
    name: torch-distributed-multi-node
  trainer:
    image: docker.io/custom-training
    command:
      - torchrun train.py
    numNodes: 5
    resourcesPerNode:
      requests:
        nvidia.com/gpu: 2
```

> The above command will be converted as follows:

Reviewer 9: ```suggestion
The above command will be converted to:
```

```bash
torchrun --nnodes=5 --nproc-per-node=2 train.py
```

> Additionally, the Kubeflow Training SDK allows to create the above `TrainJob` using the Python API:

Reviewer 9: ```suggestion
Additionally, the Kubeflow Training SDK allows the user to create the above `TrainJob` using the Python API:
```

<table>
  <tr>
   <td><strong><code>Trainer</code> Parameter</strong>
   </td>
   <td><strong> <code>TrainingRuntime</code> Parameter</strong>
   </td>
  </tr>
  <tr>
   <td><code>.image</code>
   </td>
   <td><code>.spec.replicatedJobs[name=’Node’].template.spec.template.spec.containers[name=’trainer’].image</code>
   </td>
  </tr>
  <tr>
   <td><code>.command</code>
   </td>
   <td><code>.spec.replicatedJobs[name=’Node’].template.spec.template.spec.containers[name=’trainer’].command</code>
   </td>
  </tr>
  <tr>
   <td><code>.args</code>
   </td>
   <td><code>.spec.replicatedJobs[name=’Node’].template.spec.template.spec.containers[name=’trainer’].args</code>
   </td>
  </tr>
  <tr>
   <td><code>.env</code>
   </td>
   <td><code>.spec.replicatedJobs[name=’Node’].template.spec.template.spec.containers[name=’trainer’].env</code>
   </td>
  </tr>
  <tr>
   <td><code>.numNodes</code>
   </td>
   <td><code>.spec.numNodes</code>
   </td>
  </tr>
  <tr>
   <td><code>.resourcesPerNode</code>
   </td>
   <td><code>.spec.replicatedJobs[name=’Node’].template.spec.template.spec.containers[name=’trainer’].resources</code>
   </td>
  </tr>
</table>

### The Dataset Config API

> The `DatasetConfig` represents the APIs that data scientists can use to configure dataset location.

Reviewer 9: ```suggestion
The `DatasetConfig` represents the APIs that data scientists can use to configure the dataset location.
```

	StorageUri string `json:"storageUri"`

	// Custom parameters for the dataset initializer.
	Parameters *[string]string `json:"parameters,omitempty"`

  // Reference to the secrets to access dataset.
  SecretRef corev1.SecretReference `json:"secretRef,omitempty"`
}
```

Initially we will support the following dataset providers:

- **S3:** `storageUri: s3://bucket-name/path/dataset`
- **HuggingFace:** `storageUri: hf://repo-id`

Parameters will be converted to the environment variables for the `dataset-initializer` container
in the `Initializer` Job.

For example:

```yaml
datasetConfig:
  storageUri: s3://datasets/yelp-review
  parameters:
    endpointUrl: s3.custom.com
```

Will be converted to:

```yaml
replicatedJobs:
  - name: Initializer
    template:
      spec:
        template:
          spec:
            containers:
              - name: dataset-initializer
                image: docker.io/kubeflow/dataset-initializer
                env:
                  - name: STORAGE_URI
                    value: s3://dataset/yelp-review
                  - name: ENDPOINT_URL
                    value: s3.custom.com
```

### The Model Config API

> The `ModelConfig` represents the APIs that data scientists can use to configure pre-trained model

Reviewer 9: ```suggestion
The `ModelConfig` represents the APIs that data scientists can use to configure the pre-trained model
```

type OutputModel struct {
	// Storage uri for the model exported.
	StorageUri string `json:"storageUri"`

	// Custom parameters for the model exporter.
	Parameters *[string]string `json:"parameters,omitempty"`

	// Reference to the secrets to export model.
	SecretRef corev1.SecretReference `json:"secretRef,omitempty"`
}
```

#### The Input Model API

Initially we will support the following model providers:

- **HuggingFace:** `storageUri: hf://model-name`

Parameters will be converted to the environment variables for the `model-initializer` container
in the `Initializer` Job.

For example:

```yaml
modelConfig:
  storageUri: hf://bert-based-cased
  parameters:
    transformerType: AutoModelForCausalLM
```

Will be converted to:

```yaml
replicatedJobs:
  - name: Initializer
    template:
      spec:
        template:
          spec:
            containers:
              - name: model-initializer
                image: docker.io/kubeflow/model-initializer
                env:
                  - name: STORAGE_URI
                    value: hf://bert-based-cased
                  - name: TRANSFORMER_TYPE
                    value: AutoModelForCausalLM
```

#### The Output Model API

> After initial implementation of `TrainJob` and `TrainingRuntime`, we will support ability to export

Reviewer 9: ```suggestion
After initial implementation of `TrainJob` and `TrainingRuntime`, we will support the ability to export
```

    TargetReplicatedJobs []string `json:"targetReplicatedJobs"`

    // Override parameters for Containers.
    Containers []Container `json:"container,omitempty"`

    // Override parameters for InitContainers.
    InitContainer []Container `json:"initContainer,omitempty"`

    // Override parameters for volumes.
    Volumes []corev1.Volume `json:"volume,omitempty"`

    // Custom Service Account
    ServiceAccountName string `json:"serviceAccountName,omitempty"`
}

// Override for each container.
// Parameters from Trainer, DatasetConfig, and ModelConfig will take precedence.
type Container struct {

    // Name for the container.
    Name string `json:"name"`

    // Command for the container.
    Command []string `json:"command,omitempty" protobuf:"bytes,3,rep,name=command"`

    // Args for the container.
    Args []string `json:"args,omitempty"`

    // Env for the container.
    Env []corev1.EnvVar `json:"env,omitempty"`

    // Env for the container.
    EnvFrom []corev1.EnvFromSource `json:"envFrom,omitempty"`

    // Override parameters for volume mounts.
    VolumeMounts []VolumeMount `json:"volumeMounts,omitempty"`
}
```

#### Example of TrainJob with Overrides

> This example shows how to override user-identity for sidecar container and add volume to the

Reviewer 9: ```suggestion
This example shows how to override the user-identity for the sidecar container and add volume to the
```

  trainer:
    image: docker.io/custom-training
  podSpecOverrides:
    - targetReplicatedJobs:
        - initializer
          node
      containers:
        - name: user-identity
          value: 123
        - name: trainer
          volumeMounts:
            - name: user-123-volume
              mountPath: /workspace
      volumes:
        - name: user-123-volume
          persistentVolumeClaim:
            claimName: user-123-volume
```

## The Training Runtime API

The `TrainingRuntime` is the pre-created configurations of model training on the cluster,
representing as blueprints. For example, Elastic PyTorch training, MPI DeepSpeed configuration,
BERT LLM Fine-Tuning.

These blueprints can be deployed within the Training Operator control plane and stored in a Kubeflow
public repository that users can apply to their clusters.

Platform or ML engineers can tweak existing blueprints, based on their requirements. For example,
using custom configurations.

The Kubeflow Training Operator can maintain more Training Runtimes when the community is ready to
support them. For example, runtimes for [Jax](https://jax.readthedocs.io/en/latest/index.html) or

> [MLX](https://ml-explore.github.io/mlx/build/html/index.html). Initially, we will support: PyTorch,

Reviewer: The non goal states that we will not support Tensorflow, XGBoost and PaddlePaddle.

Author: I can add it to the non-goals, but after initial implementation with PyTorch we are planning to add those runtimes.
WDYT @kubeflow/wg-training-leads ?

Reviewer: ### Non-Goals

- Support MPI V1 implementation.
- Distributed training for TensorFlow, XGboost, JAX, and PaddlePaddle will be added after initial
  implementation.
- Migrate Kubeflow V1 controller to use `JobSet`.

I just mean you mention it as a nongoal already.

Author: Oh, yeah. Let me just add the message that these runtimes will be added after initial implementation.

MPI, TensorFlow, XGBoost, and PaddlePaddle.

The `TrainingRuntime` is immutable, and so to make a change, a new version of the `TrainingRuntime`

> must be created and then changing the `TrainJob` to point to the new version.

Reviewer 9: ```suggestion
must be created and then the user must change the `TrainJob` to point to the new version.
```

We are going to create two CRDs: `TrainingRuntime` and `ClusterTrainingRuntime`. These runtimes have
exactly the same APIs, but the first one is the namespace-scoped, the second is the cluster-scoped.
If `trainingRuntimeRef` in `TrainJob` has the namespace, controller will use the `TrainingRuntime`,
otherwise it will use the `ClusterTrainingRuntime`.

```golang
type TrainingRuntime struct {
    metav1.TypeMeta   `json:",inline"`
    metav1.ObjectMeta `json:"metadata,omitempty"`

    // Framework specific parameters.
    MLSpec *MLSpec `json:"mlSpec,omitempty"`

    // Number of nodes to execute training.
    NumNodes int `json:"numNodes,omitempty"`

    // JobSet spec.
    JobSetSpec *batchv1.JobSetSpec `json:",inline"`

    // For gang-scheduling using volcano or scheduler plugins, supported for all frameworks.
    GangScheduler *GangScheduler `json:"gangScheduler,omitempty"`
}

// One of the specs can be selected.
type MLSpec struct {

    // Custom Spec for Torch
    TorchSpec *TorchSpec `json:"torchSpec,omitempty"`

    // Custom Spec for MPI
    MPISpec *MPISpec `json:"mpiSpec,omitempty"`
}
```

### The Gang Scheduler API

> Gang scheduler plugin is used to create appropriate `PodGroup` for Volcano or scheduler plugins.

Reviewer 9: ```suggestion
Gang scheduler plugin is used to create the appropriate `PodGroup` for Volcano or scheduler plugins.
```

Single worker training:

```yaml
apiVersion: kubeflow.org/v2alpha1
kind: ClusterTrainingRuntime
metadata:
  name: torch-simple
spec:
  replicatedJobs:
    - name: node
      template:
        spec:
          template:
            spec:
              containers:
                - name: trainer
                  image: docker.io/kubeflow/pytorch-mnist
                  command:
                    - torchrun train.py
```

Single node multi worker training:

```yaml
apiVersion: kubeflow.org/v2alpha1
kind: ClusterTrainingRuntime
metadata:
  name: torch-distributed-single-worker
spec:
  mlSpec:
    torch:
      numProcPerNode: 5
      standalone: True
  replicatedJobs:
    - name: Node
      template:
        spec:
          template:
            spec:
              containers:
                - name: trainer
                  image: docker.io/kubeflow/pytorch-mnist
                  env:
                    - name: MASTER_ADDR
                      value: "pytorch-node-0-0.pytorch"
                    - name: MASTER_PORT
                      value: 29400
                  command:
                    - torchrun train.py
```

#### LLM Fine-Tuning Runtimes

> In the future, we can consider to use [the `torchtune` CLI](https://github.com/pytorch/torchtune/tree/main)

Reviewer 9: ```suggestion
In the future, we can consider using [the `torchtune` CLI](https://github.com/pytorch/torchtune/tree/main)
```

for Fine-Tuning with PyTorch.

##### Llama 7b

> The following runtime can be used for Llama 7b model.

Reviewer 9: ```suggestion
The following runtime can be used for the Llama 7b model.
```

                  env:
                    - name: MASTER_ADDR
                      value: "pytorch-node-0-0.pytorch"
                    - name: MASTER_PORT
                      value: 29400
                    - name: TRANSFORMER_TYPE
                      value: AutoModelForCausalLM
                    - name: LORA_CONFIG
                      value: |
                        {"peft_type": "LORA", "r": 8, "lora_alpha": 16}
                  command:
                    - torchrun hf_llm_training.py
                  resources:
                    limits:
                      nvidia.com/gpu: 2
                  volumeMounts:
                    - mountPath: /workspace/dataset
                      name: dataset-initializer
                    - mountPath: /workspace/model
                      name: model-initializer
              volumes:
                - name: dataset-initializer
                  persistentVolumeClaim:
                    claimName: dataset-initializer
                - name: model-initializer
                  persistentVolumeClaim:
                    claimName: model-initializer
```

#### MPI Runtime

> For MPI, we can add support the `DeepSpeed` runtimes.

Reviewer 9: ```suggestion
For MPI, we can add support for the `DeepSpeed` runtimes.
```

kind: ClusterTrainingRuntime
metadata:
  name: mpi-simple
spec:
  mlSpec:
    mpi:
      mpiImplementation: OpenMPI
      numProcPerNode: 5
  numNodes: 5
  replicatedJobs:
    - name: Launcher
      template:
        spec:
          template:
            spec:
              containers:
                - name: mpi-launcher
                  image: docker.io/mpi-launch
                  command:
                    - mpirun -np 5 --host mpi-simple.default.svc
    - name: Node
      template:
        spec:
          template:
            spec:
              containers:
                - name: trainer
                  image: docker.io/mpi-training
                  command:
                    - mpirun -np 2 train.py
```

#### TensorFlow Runtime

_Will be added after initial implementation for PyTorch._

#### XGBoost Runtime

_Will be added after initial implementation for PyTorch._

#### Paddle Runtime

_Will be added after initial implementation for PyTorch._

#### Jax Runtime

_Will be added after initial implementation for PyTorch._

## Migration from Kubeflow Training V1

These API changes will not be compatible with Training Operator V1 APIs. Thus, existing users have

> to migrate to the newer APIs. Kubeflow community will provide instructions on how to migrate existing

Reviewer 9: ```suggestion
to migrate to the newer APIs. The Kubeflow community will provide instructions on how to migrate existing
```

and built [a new Kubernetes SIGs project: `JobSet`](https://github.com/kubernetes-sigs/jobset) to
manage groups of `Jobs`.

This document consolidates efforts for the Cloud Native ML Training between Kubeflow and Kubernetes
communities.

## Motivation

We often implement features similar to batch/v1 `Job`, such as “suspend”, on the Training Operator
side since the Training Operator creates blocks of plain Pod and Service for each rank once
Kubeflow Jobs are created. However, if we continue taking the same approach to use lowest level
abstractions that introduce redundancy, the maintenance costs will continue to increase.

Replacing repetitive infrastructure layers with `JobSet` would help to avoid redundancy and reduce
developer toil.

Additionally, introducing `JobSet` as an infrastructure layer would allow us to introduce batch
workload features such as
[the PodFailurePolicy](https://kubernetes.io/docs/concepts/workloads/controllers/job/#pod-failure-policy)
and [the PodDisruptionCondition](https://kubernetes.io/docs/concepts/workloads/controllers/job/#handling-pod-and-container-failures)
easily.

> Please also see the [Kubernetes JobSet and Kubeflow Training Operator collaboration document](https://docs.google.com/document/d/1C2ev7yRbnMTlQWbQCfX7BCCHcLIAGW2MP9f7YeKl2Ck/edit?usp=sharing).

Reviewer 9: This is great! And it's really helpful context. I briefly reviewed the JobSet docs / concepts and it's intuitive why we would want to move fault tolerance into a separate project. 

In previous roles, along with having to deal with pod/job failures, I've also had to spend a lot of time on dealing with capacity constraints/issues especially among different teams. Some folks at IBM recently shared with me work they've done and I'm curious to see where capacity management would be best fit for large-scale model training? 

Given the separation of pod failures into JobSet, I don't think it makes sense for capacity management to go in the Training Operator but would a separate project make more sense then?

Would love any thoughts here.

Author: > I'm curious to see where capacity management would be best fit for large-scale model training?

I think, we should discuss it with Batch WG (cc @tenzen-y @alculquicondor @ahg-g @kannon92 @danielvegamyhre). Especially the use-cases and what kind of capacity management we are talking about.

Reviewer: Yes please. Defining "capacity issues" would be helpful but we should discuss outside of this KEP.

Reviewer 9: Sounds good! I'll add it to the agenda.

          value: 123
        - name: trainer
          volumeMounts:
            - name: user-123-volume
              mountPath: /workspace
      volumes:
        - name: user-123-volume
          persistentVolumeClaim:
            claimName: user-123-volume
```

## The Training Runtime API

The `TrainingRuntime` is the pre-created configurations of model training on the cluster,
representing as blueprints. For example, Elastic PyTorch training, MPI DeepSpeed configuration,
BERT LLM Fine-Tuning.

These blueprints can be deployed within the Training Operator control plane and stored in a Kubeflow
public repository that users can apply to their clusters.

Platform or ML engineers can tweak existing blueprints, based on their requirements. For example,
using custom configurations.

The Kubeflow Training Operator can maintain more Training Runtimes when the community is ready to
support them. For example, runtimes for [Jax](https://jax.readthedocs.io/en/latest/index.html) or
[MLX](https://ml-explore.github.io/mlx/build/html/index.html). We will support PyTorch and MPI runtimes.
After initial implementation, we will support TensorFlow, XGboost, and PaddlePaddle runtimes, but

> it is out of scope of this KEP.

Reviewer: ```suggestion
it is out of scope for this KEP.
```

> The `TrainingRuntime` is immutable, and so to make a change, a new version of the `TrainingRuntime`

Reviewer: @tenzen-y How would immutability impact Kueue integration?

I'm trying to think through what kind of edits kueue or other controllers would want to add to TrainJob.

I know that for JobSet we had to relax quite a few immutability requirements for Kueue integration so I think its worth thinking through this.

Reviewer 5: This immutable is affected only on the `TrainingRuntime` and `ClusterTrainingRuntime`. So, we can still modify the TrainJob. Hence, I think that this will not bring us any specific confusion in JobSet and Kueue.

@kannon92 If I am missing anything, please let me know.

Reviewer: For example, in JobSet, we had to relax some of these rules:

https://github.com/kubernetes-sigs/jobset/blob/main/pkg/webhooks/jobset_webhook.go#L320

I just want to make sure that integration for Kueue has everything we need in TrainJob rather than TrainingRuntime. It sounds like Platform engineers define the TrainingRuntime and that template is reused for TrainJob.

But do we need to update anything in TrainingRuntime for Kueue?

Reviewer 5: I imagined that Kueue modifies the PodOverrideSpec field: https://github.com/kubeflow/training-operator/blob/02947981c995e3aec0235bef28bd7fc2c6aa5a1a/docs/proposals/2170-kubeflow-training-v2/README.md#the-pod-spec-overrides-apis

But, it seems that the podSpecOverride doesn't have scheduling directive. We need to allow external controller to modify the scheduling directive so that Kueue can handle the objects.
Thanks for a great call!

@andreyvelich Could you add scheduling directives to the podSpecOverrides so that we can natively support Kueue?

- nodeSelectors
- tolerations
- schedulingGates

@kannon92 If I'm missing any scheduling directives, please let me know.

Reviewer 5: Oh, additionally, the TrainJob does not expose to modify the managedBy and suspend field.

Reviewer: Thank you for checking that out. I have no more concerns. LGTM.

Author: How will the TrainJob scheduling directives will work together with JobSet ? @tenzen-y Let's discuss it in more details tomorrow, so we can see how we should introduce those APIs.

Reviewer 5: Sure.

# KEP-2170: Kubeflow Training V2 API

## Authors

- Andrey Velichkevich - [@andreyvelich](https://github.com/andreyvelich)
- Yuki Iwai - [@tenzen-y](https://github.com/tenzen-y)

Creation date: 2024-07-16

Google doc: https://bit.ly/3WzjTlw

## Overview

This document discusses the new Kubeflow Training V2 API.

When we built the
[Kubeflow Training Operator a couple of years ago](https://docs.google.com/document/d/1x1JPDQfDMIbnoQRftDH1IzGU0qvHGSU4W6Jl4rJLPhI/edit?usp=sharing),
Kubernetes lacked better features to support distributed machine learning (ML) training, such as

> SuccessPolicy and RestartPolicy (FailurePolicy). Recently, the Kubernetes community launched the

Reviewer 5: Actually, this RestartPolicy indicates the Kubeflow RestartPolicy :)
https://github.com/kubeflow/training-operator/blob/736c8147bbd9efbdd273bc8e3665d11c90294c8f/pkg/apis/kubeflow.org/v1/common_types.go#L100-L103

> SuccessPolicy and RestartPolicy (FailurePolicy). Recently, the Kubernetes community launched the

Reviewer 10: I see, so maybe give links to disambiguate the concepts cause it wasn't clear for me

> SuccessPolicy and RestartPolicy (FailurePolicy). Recently, the Kubernetes community launched the

Author: I clarified this part and added links.

1. **Data Scientists/ML Engineers**. They create model architectures and advanced ML algorithms to train models.
   They prefer to use Python for their work. They are aware of `torch.nn` APIs, but not with
   `torch.distributed` and Kubernetes concepts to scale model training.

Based on the above personas, we should build an API that everyone will benefit from.

### Goals

- Introduce the `TrainingRuntime` and `ClusterTrainingRuntime` APIs that will store blueprints
  for model training and LLM fine-tuning using various ML frameworks. These runtimes will be built
  on top of `JobSet` APIs with additional functionality for special use-cases.
  For example, training using MPI orchestration.
- Introduce Kubeflow `TrainJob` API that allows to reuse these runtimes and quickly start a new
  training job without understanding complex Kubernetes APIs.
- Update Kubeflow Training SDK to allow data scientists quickly create and monitor `TrainJobs`.
- Create community-supported `ClusterTrainingRuntime` for distributed training with PyTorch and MPI.
- Create community-supported `ClusterTrainingRuntime` for LLM fine-tuning for various foundational
  models (e.g. Mistral, LLama-70b, Gemma-7b).

> - Work on the following `JobSet` improvements: https://github.com/kubernetes-sigs/jobset/issues/463

Reviewer 5: Additionally, we need to work on https://github.com/kubernetes-sigs/jobset/issues/570 so that we can netively support the Elastic PyTorchJob.

  and https://github.com/kubernetes-sigs/jobset/issues/572
- Integrate `TrainJob` with Kueue and MultiKueue to effectively manage resources for training jobs
  and orchestrate resources across multiple clusters.

### Non-Goals

- Support MPI V1 implementation.
- Distributed training for TensorFlow, XGboost, JAX, and PaddlePaddle will be added after initial
  implementation.
- Migrate Kubeflow V1 controller to use `JobSet`.

> 

Reviewer 5: ```suggestion
- Propose the migration mechanisms / ways from Kubeflow Training v1 to v2.
```

How about to mention migration ways here? I guess that providing the migration ways like conversion webhooks not easy since there is not any compatibility between TrainJob and TrainingRuntime APIs.

I guess that dedicated KEP for the migration ways would be worth it.

Author: Let me say that we will create migration plan in the future.

> 

Reviewer 5: ```suggestion
- Propose the Kubeflow Train Python SDK v2 API design
```

It seems that this proposal doesn't have a specific Python API design, although this proposal includes the example idea for the Python SDK v2.

Author: Yes, let me add it to the non-goal.

for other MPI implementations: Intel MPI, MPICH, Spectrum MPI.

## The TrainJob API

The `TrainJob` exposes APIs that data scientist can override in `TrainingRuntime` to create
a training job:

```golang
type TrainJob struct {
	metav1.TypeMeta   `json:",inline"`
	metav1.ObjectMeta `json:"metadata,omitempty"`

	// Spec defines the desired state of TrainJob.
	Spec TrainJobSpec `json:"spec"`

	// Status defines the current state of TrainJob.
	Status TrainJobStatus `json:"status,omitempty"`
}

type TrainJobSpec struct {
	// Reference to the Training Runtime.
	TrainingRuntimeRef *TrainingRuntimeRef `json:"trainingRuntimeRef"`

	// Parameters that data scientists can override
	Trainer *Trainer `json:"trainer,omitempty"`

	// Configuration for training dataset
	DatasetConfig *DatasetConfig `json:"datasetConfig,omitempty"`

	// Configuration for the pre-trained model and location for model output
	ModelConfig *ModelConfig `json:"modelConfig,omitempty"`

	// Custom metadata to apply for Job, JobSet, etc.
	Labels      map[string]string `json:"labels,omitempty"`
	Annotations map[string]string `json:"annotations,omitempty"`

> }

Reviewer 5: ```suggestion

	// podSpecOverrides represents overrides for the TrainingRuntime when TrainJob is created.
	PodSpecOverrides []PodSpecOverrides `json:"podSpecOverrides,omitempty"`
}
```

Author: Nice catch!

type TrainingRuntimeRef struct {
	// Name for the training runtime.
	Name string `json:"name"`
	// Namespace for the runtime.
	// If namespace is set, TrainingRuntime is used. Otherwise, ClusterTrainingRuntime is used.
	Namespace string `json:"namespace,omitempty"`

> }

Reviewer 5: As we discussed in Google Doc and previous meeting, could we add the Group/Version/Kind since this switching mechanism depending on the `namespace` field indicates the implied the specifications. It would be better to define all specifications as APIs.

Author: Should we initially just introduce Kind since we don't support the runtime versioning ?
Also, are we going to validate that if `Kind: ClusterTrainingRuntime` is set, then `namespace` parameter should be ignored ?

Reviewer 5: Can we add the Kind and Version? In this version doesn't mean the runtime revision, which indicates the API version like v2alpha1.

```yaml
trainingRuntimeRef:
  name: torch-multi-node
  namespace: tenant-a  # When they use the TrainingRuntime, they need to specify this field.
  kind: TrainingRuntime | ClusterTrainingRuntime
  version: v2alpha1 # This is API version
  revision: v1 # This will be supported in the future after we introduce the runtime revision management mechanism.
```

Author: Should we re-use: https://github.com/kubernetes/api/blob/master/core/v1/types.go#L585C6-L585C26 ?
Or we don't need the `group` since it is always `kubeflow.org`?

   </td>
  </tr>
  <tr>
   <td><code>.command</code>
   </td>
   <td><code>.spec.replicatedJobs[name=’Node’].template.spec.template.spec.containers[name=’trainer’].command</code>
   </td>
  </tr>
  <tr>
   <td><code>.args</code>
   </td>
   <td><code>.spec.replicatedJobs[name=’Node’].template.spec.template.spec.containers[name=’trainer’].args</code>
   </td>
  </tr>
  <tr>
   <td><code>.env</code>
   </td>
   <td><code>.spec.replicatedJobs[name=’Node’].template.spec.template.spec.containers[name=’trainer’].env</code>
   </td>
  </tr>
  <tr>
   <td><code>.numNodes</code>
   </td>
   <td><code>.spec.numNodes</code>
   </td>
  </tr>
  <tr>
   <td><code>.resourcesPerNode</code>
   </td>
   <td><code>.spec.replicatedJobs[name=’Node’].template.spec.template.spec.containers[name=’trainer’].resources</code>
   </td>
  </tr>
</table>

### The Dataset Config API

The `DatasetConfig` represents the APIs that data scientists can use to configure the dataset location.

```golang
type DatasetConfig struct {

	// Storage uri for the dataset provider.
	StorageUri string `json:"storageUri"`

	// Custom parameters for the dataset initializer.

> 	Parameters *[string]string `json:"parameters,omitempty"`

Reviewer 5: ```suggestion
	Parameters map[string]string `json:"parameters,omitempty"`
```

Reviewer 5: I still wondering if we can have scheme safe union type parameters. Because this map parameters do not allow us to add any CRD based validation and so on. I'm suspecting that this non scheme safe parameters API could be blocker for introducing the CRD validations.

But, I'm ok with this non scheme safe parameters for the alpha API. Before we graduate API to Beta, we can revisit here, and we can consider which parameters we should define as dedicated field.

Reviewer 5: Basically, I agree with parameter fields in alpha API since it's challenging to decide which model/storage provider parameters should be. defined in the API during the first implementation. So, I would like to revisit here after first implementation and collected feedback from users.

Author: Sounds good!

>   // Reference to the secrets to access dataset.
>   SecretRef corev1.SecretReference `json:"secretRef,omitempty"`

Reviewer 5: ```suggestion
	// Reference to the secrets to access dataset.
	SecretRef corev1.SecretReference `json:"secretRef,omitempty"`
```

in the `Initializer` Job.

For example:

```yaml
datasetConfig:
  storageUri: s3://datasets/yelp-review
  parameters:
    endpointUrl: s3.custom.com
```

Will be converted to:

```yaml
replicatedJobs:
  - name: Initializer
    template:
      spec:
        template:
          spec:
            containers:
              - name: dataset-initializer
                image: docker.io/kubeflow/dataset-initializer
                env:
                  - name: STORAGE_URI
                    value: s3://dataset/yelp-review
                  - name: ENDPOINT_URL
                    value: s3.custom.com
```

### The Model Config API

The `ModelConfig` represents the APIs that data scientists can use to configure the pre-trained model
input and output location.

```golang
type ModelConfig struct {
	// Configuration for pre-trained model.
	Input *InputModel `json:"input,omitempty"`

	// Configuration for trained model.
	Output *OutputModel `json:"output,omitempty"`
}

type InputModel struct {
	// Storage uri for the model provider.
	StorageUri string `json:"storageUri"`

	// Custom parameters for the model initializer.

> 	Parameters *[string]string `json:"parameters,omitempty"`

Reviewer 5: ```suggestion
	Parameters map[string]string `json:"parameters,omitempty"`
```

	// Reference to the secrets to access model.
	SecretRef corev1.SecretReference `json:"secretRef,omitempty"`
}

type OutputModel struct {
	// Storage uri for the model exported.
	StorageUri string `json:"storageUri"`

	// Custom parameters for the model exporter.

> 	Parameters *[string]string `json:"parameters,omitempty"`

Reviewer 5: ```suggestion
	Parameters map[string]string `json:"parameters,omitempty"`
```

These parameters can include the user's identity or PVC.

Usually, these parameters should not be configured by the user and should be attached during the
orchestration (e.g. using Kubernetes admission webhooks or custom clients).

In the future, we can add more parameters if we find use-cases when it is required.

```golang
type PodSpecOverride struct {
    // Name of the training replica in the training runtime template to override
    TargetReplicatedJobs []string `json:"targetReplicatedJobs"`

    // Override parameters for Containers.
    Containers []Container `json:"container,omitempty"`

    // Override parameters for InitContainers.
    InitContainer []Container `json:"initContainer,omitempty"`

    // Override parameters for volumes.
    Volumes []corev1.Volume `json:"volume,omitempty"`

    // Custom Service Account
    ServiceAccountName string `json:"serviceAccountName,omitempty"`
}

// Override for each container.
// Parameters from Trainer, DatasetConfig, and ModelConfig will take precedence.
type Container struct {

    // Name for the container.
    Name string `json:"name"`

    // Command for the container.

>     Command []string `json:"command,omitempty" protobuf:"bytes,3,rep,name=command"`

Reviewer 5: ```suggestion
    Command []string `json:"command,omitempty"`
```

    // Args for the container.
    Args []string `json:"args,omitempty"`

    // Env for the container.
    Env []corev1.EnvVar `json:"env,omitempty"`

    // Env for the container.
    EnvFrom []corev1.EnvFromSource `json:"envFrom,omitempty"`

    // Override parameters for volume mounts.
    VolumeMounts []VolumeMount `json:"volumeMounts,omitempty"`
}
```

#### Example of TrainJob with Overrides

This example shows how to override the user-identity for the sidecar container and add volume to the
trainer container.

```yaml
apiVersion: kubeflow.org/v2alpha1
kind: TrainJob
metadata:
  name: pytorch-distributed
  namespace: tenant-alpha
spec:
  trainingRuntimeRef:
    name: pytorch-distributed-gpu
  trainer:
    image: docker.io/custom-training
  podSpecOverrides:
    - targetReplicatedJobs:
        - initializer

>           node

Reviewer 5: ```suggestion
```

Author: I think, we should remove initializer from here.

must be created and then the user must change the `TrainJob` to point to the new version.
This provides control as to how changes to runtimes propagate to existing training jobs.
For example, when training is running for a long time (e.g. 1-2 months).

In the future implementation, we will introduce a revision control mechanism similar to
[Kubernetes Deployment](https://kubernetes.io/docs/concepts/workloads/controllers/deployment/#updating-a-deployment)
to control versions of `TrainingRuntime` and enable rolling updates.

We are going to create two CRDs: `TrainingRuntime` and `ClusterTrainingRuntime`. These runtimes have
exactly the same APIs, but the first one is the namespace-scoped, the second is the cluster-scoped.
If `trainingRuntimeRef` in `TrainJob` has the namespace, controller will use the `TrainingRuntime`,
otherwise it will use the `ClusterTrainingRuntime`.

```golang
type TrainingRuntime struct {
    metav1.TypeMeta   `json:",inline"`
    metav1.ObjectMeta `json:"metadata,omitempty"`

    // Framework specific parameters.
    MLSpec *MLSpec `json:"mlSpec,omitempty"`

    // Number of nodes to execute training.

>     NumNodes int `json:"numNodes,omitempty"`

Reviewer 5: ```suggestion
    NumNodes *int32 `json:"numNodes,omitempty"`
```

This is optional field, right?

Author: Actually, we should set it to 1 by default, isn't ?

Reviewer 5: > we should set it to 1 by default, isn't ?

That's right. If users do not set this field (numNodes == nil), we can default to 1.

    // JobSet spec.
    JobSetSpec *batchv1.JobSetSpec `json:",inline"`

    // For gang-scheduling using volcano or scheduler plugins, supported for all frameworks.
    GangScheduler *GangScheduler `json:"gangScheduler,omitempty"`
}

// One of the specs can be selected.
type MLSpec struct {

    // Custom Spec for Torch
    TorchSpec *TorchSpec `json:"torchSpec,omitempty"`

    // Custom Spec for MPI
    MPISpec *MPISpec `json:"mpiSpec,omitempty"`
}
```

### The Gang Scheduler API

Gang scheduler plugin is used to create the appropriate `PodGroup` for Volcano or scheduler plugins.

```golang
type GangScheduler struct {
    // Plugin for gang scheduling.
    Plugin *GangSchedulerPlugin `json:plugin,omitempty"`

    // Time threshold to schedule PodGroup for gang scheduling.

>     ScheduleTimeoutSeconds string `json:scheduleTimeoutSeconds,omitempty"`

Reviewer 5: ```suggestion
    ScheduleTimeoutSeconds *string `json:scheduleTimeoutSeconds,omitempty"`
```

Let's define the omitempty field as a pointer.

}

type GangSchedulerPlugin string

const (
    GangSchedulerPluginVolcano GangSchedulerPlugin = "volcano"

>     GangSchedulerPlugins       GangSchedulerPlugin = "scheduler-plugins"

Reviewer 5: ```suggestion
    SchedulerPlugins       GangSchedulerPlugin = "scheduler-plugins"
```

Actually, this is the coscheduling scheduler-plugins.

Author: If we initially support only Coscheduling: https://github.com/kubernetes-sigs/scheduler-plugins?tab=readme-ov-file#plugins, should we call it: `GangSchedulerPluginCoscheduling` ?

Reviewer 5: Indeed, we support only the coscheduling plugin. I'm fine with either name.

Note that the scheduler-plugins support the gang scheduling only by the coscheduling plugin.

Author: Makes sense, I think it is better to be more declarative here.

)
```

### The Torch Spec API

The `TorchSpec` API represents the configuration for the PyTorch distributed training. This configuration
allows platform engineers to explicitly configure `torchrun` setting.

The distributed parameters are taken from the
[PyTorch distributed launch run](https://github.com/pytorch/pytorch/blob/94dc3253a0fefbfb95fbe467ddd628e4c2eb08d7/torch/distributed/run.py).

For Elastic Training we will always pass the following parameters:

```bash
--rdzv-backend=c10d

--rdzv-id will be set automatically.

--rdzv-endpoint will always point to the node-0 Pod.
```

Since the [etcd and etcd-v2 are legacy rendezvous](https://pytorch.org/docs/stable/elastic/run.html#note-on-rendezvous-backend),
we won't support them in `TorchSpec`. We can introduce them in the future if users will require them.

```golang
// TorchSpec represents the configuration for PyTorch.
type TorchSpec struct {

    // Number of Procs per Node.

>     NumProcPerNode int `json:"numProcPerNode,omitempty"`

Reviewer 5: ```suggestion
    NumProcPerNode *int32 `json:"numProcPerNode,omitempty"`
```

    // Used for single-node multi-worker training

>     Standalone bool `json:"standalone,omitempty"`

Reviewer 5: ```suggestion
    Standalone bool `json:"standalone"`
```

Let's handle this bool field as non omitempty field so that we can set false as a default.

Author: @tenzen-y In that case do we need to configure: `standalone: False` in every Torch runtime ?

Reviewer 5: The answer is no. Users can skip to specify the `standalone: false` when they do not want to use the standalone mode.

Author: Since this can be omitted in the TrainingRuntime YAML, why we don't want to set it as:
```go
 Standalone bool `json:"standalone,omitempty"`
 ```

// --nnodes=minNodes:maxNodes
type TorchElasticPolicy struct {

    // The limits to restart TrainJob.
    // Insert it to the JobSet.spec.failurePolicy.maxRestarts
    MaxRestarts *in32 `json:"maxRestarts,omitempty"`

    // Min number of nodes for HPA and torchrun.
    MinNodes *in32 `json:"minNodes,omitempty"`

    // Max number of nodes for HPA and torchrun.
    MaxNodes *in32 `json:"maxNodes,omitempty"`

    // Metrics for scale up and down replicas.
    Metrics []autoscalingv2.MetricSpec `json:"metrics,omitempty"`
}
```

### The MPI Spec API

The `MPISpec` API represents the configuration for training using MPI orchestration.
E.g. creation of host-files and SSH keys. Using MPI might be more efficient for training on HPC
clusters or for some ML frameworks (e.g. [MLX distributed with MPI](https://ml-explore.github.io/mlx/build/html/usage/distributed.html)).

We will fully migrate to the MPI Operator V2 functionality as part of this KEP.
Check [the proposal for the MPI V2 APIs.](https://github.com/kubeflow/mpi-operator/blob/master/proposals/scalable-robust-operator.md)

```golang
type MPISpec struct {
    // Number of Procs per Node.

>     NumProcPerNode int `json:"numProcPerNode,omitempty"`

Reviewer 5: ```suggestion
    NumProcPerNode *int32 `json:"numProcPerNode,omitempty"`
```

    // MPI Implementation to create appropriate host-files.
    // Can be one of OpenMPI, Intel, or MPICH.

>     MPIImplementation MPIImplementation `json:"mpiImplementation,omitempty"`

Reviewer 5: ```suggestion
    MPIImplementation MPIImplementation `json:"mpiImplementation"`
```

Let's use this field as a necessary field and then default to OpenMPI when users do not specify this field.

    // Directory where SSH keys are mounted.

>     SSHAuthMountPath string `json:"SSHAuthMountPath,omitempty"`

Reviewer 5: ```suggestion
    SSHAuthMountPath *string `json:"SSHAuthMountPath,omitempty"`
```

> }

Reviewer 5: Could we support the `runLauncherAsWorker` API the same as the v2 MPIJob API so that we can perform the training task in the rank 0 as well?

Author: Should we collect a feedback about this feature ?
Don't you think that creating 2 separate runtimes will be sufficient:
- When Launcher is not running training
- When Launcher is running training

Reviewer 5: When we enable runLauncherAsWorker, we need to join the launcher Node to the headless service. So, we need to have the dedicated API.

Author: @tenzen-y In that case, will Launcher and Node have different JobSet templates ?

Reviewer 5: @andreyvelich Yes, the launcher executes the `mpirun` command, and the workers execute the ssh server.

Note that it should not be better approach to predict the launcher mode based on the number of jobTemplates since it indicates that the controller logics have any implied specifications. In general, exposing the specifications as APIs should be better.

   </td>
   <td><code>python</code> 
   </td>
   <td><a href="https://github.com/dmlc/xgboost/blob/a5a58102e5e82fa508514c34cd8e5f408dcfd3e1/python-package/xgboost/tracker.py#L17">Rabit Tracker</a> for c10d
   </td>
  </tr>
  <tr>
   <td>DeepSpeed
   </td>
   <td>Slot
   </td>
   <td>Node
<p>
<code>(--num_nodes)</code>
   </td>
   <td><code>deepspeed</code>
   </td>
   <td><a href="https://www.deepspeed.ai/getting-started/#resource-configuration-multi-node">DeepSpeed Distributed</a>
   </td>
  </tr>
</table>

Additionally, check [this document for the `mpirun` command](https://gist.github.com/vsoch/9ac7c4448dffe656d946edceaa58bd9e)
for other MPI implementations: Intel MPI, MPICH, Spectrum MPI.

## The TrainJob API

The `TrainJob` exposes APIs that data scientist can override in `TrainingRuntime` to create
a training job:

```golang
type TrainJob struct {
	metav1.TypeMeta   `json:",inline"`
	metav1.ObjectMeta `json:"metadata,omitempty"`

	// Spec defines the desired state of TrainJob.
	Spec TrainJobSpec `json:"spec"`

	// Status defines the current state of TrainJob.
	Status TrainJobStatus `json:"status,omitempty"`
}

type TrainJobSpec struct {
	// Reference to the Training Runtime.

> 	TrainingRuntimeRef *TrainingRuntimeRef `json:"trainingRuntimeRef"`

Reviewer 5: ```suggestion
	TrainingRuntimeRef TrainingRuntimeRef `json:"trainingRuntimeRef"`
```

This is a necessary field, right?

   </td>
   <td>Configuration for the dataset that will be used to train or fine-tune model
   </td>
  </tr>
  <tr>
   <td>Labels and Annotations
   </td>
   <td>Custom metadata that needs to be applied to the <code>TrainJob</code> resources: JobSet, Job, and Pods.
   </td>
  </tr>
  <tr>
   <td><code>PodSpecOverrides</code>
   </td>
   <td>Custom overrides that are specific to the <code>TrainJob</code> and need to be applied to the
    <code>TrainJob</code> resources. For example, the user identity. Usually, it is managed by
    custom admission webhooks that inject data to the <code>TrainJob</code> after the user creates it
    via the Python SDK or <code>kubectl</code>
   </td>
  </tr>
  <tr>
   <td>Suspend and ManagedBy
   </td>
   <td>Scheduling directives for Kueue and MultiKueue
   </td>
  </tr>
</table>

### Example of TrainJob

```yaml
apiVersion: kubeflow.org/v2alpha1
kind: TrainJob
metadata:
  name: torch-ddp
  namespace: tenant-alpha
spec:
  trainingRuntimeRef:
    name: torch-distributed-multi-node
  trainer:
    image: docker.io/custom-training
    command:
      - torchrun train.py
    numNodes: 5
    resourcesPerNode:
      requests:
        nvidia.com/gpu: 2
```

> The above command will be converted to:

Author: Nice catch, let me clarify this

scheduler is not the same as `PodGroup` plugin.

```yaml
apiVersion: kubeflow.org/v2alpha1
kind: ClusterTrainingRuntime
metadata:
  name: torch-distributed-multi-node
spec:
  mlSpec:
    torch:
      numProcPerNode: 5
  podGroupSpec:
    plugin: coscheduling
    scheduleTimeoutSeconds: 100
  replicatedJobs:
    - name: node
      template:
        spec:
          template:
            spec:
              schedulerName: coscheduling
              containers:
                - name: trainer
                  image: docker.io/kubeflow/pytorch-mnist
                  resources:
                    limits:
                      nvidia.com/gpu: 1
                  env:
                    - name: MASTER_ADDR
                      value: "pytorch-node-0-0.pytorch"
                    - name: MASTER_PORT
                      value: 29400
                  command:
                    - torchrun train.py
```

Training Operator will create the `PodGroup` using the following spec:

```yaml
apiVersion: scheduling.x-k8s.io/v1alpha1
kind: PodGroup
metadata:
  name: nginx
spec:
  scheduleTimeoutSeconds: 100
  minMember: 5

> ```

Reviewer 5: That is a great PodGroup specification unveil!

)
```

Here is the example of runtime with gang-scheduling using coscheduling plugin.

```yaml
apiVersion: kubeflow.org/v2alpha1
kind: ClusterTrainingRuntime
metadata:
  name: torch-distributed-multi-node
spec:
  mlSpec:
    torch:
      numProcPerNode: 5
  podGroupSpec:
    plugin: coscheduling
    scheduleTimeoutSeconds: 100
  replicatedJobs:
    - name: node
      template:
        spec:
          template:
            spec:
              schedulerName: coscheduling
              containers:
                - name: trainer
                  image: docker.io/kubeflow/pytorch-mnist
                  resources:
                    limits:
                      nvidia.com/gpu: 1
                  env:
                    - name: MASTER_ADDR
                      value: "pytorch-node-0-0.pytorch"
                    - name: MASTER_PORT
                      value: 29400
                  command:
                    - torchrun train.py
```

> Training Operator will create the `PodGroup` using the following spec:

Reviewer 3: This makes sense. Naive question - does this imply that any custom gang scheduler used is required to make a PodGroup to function? Or if you have a plugin that doesn't create a PodGroup (and does its own thing) the group would be made anyway? And from the example above, how is `minMember` of 5 derived? You would probably need custom logic per scheduler backend for this API. For example, the PodGroup created by coscheduling vs volcano.sh are different underlying objects. (https://github.com/kubernetes-sigs/scheduler-plugins/blob/d67d2c15199595ccc6218574bd8e07b68b7146f4/apis/scheduling/v1alpha1/types.go#L128  and https://github.com/volcano-sh/apis/blob/78d912ce096c9fd9120488b5c7b999d4996f2cb5/pkg/apis/scheduling/types.go#L147)

Reviewer 3: If this Training API is going t take on the challenge of orchestrating the logic for creation of needed PodGroup (varying by the plugin) then this probably works OK. Otherwise, I'd still place the responsibility on the user for now for creating the pod group, but just allow the `schedulerName` and other labels to use it in whatever turns into the actual job or podspec.

Author: >does this imply that any custom gang scheduler used is required to make a PodGroup to function? 

The PodGroup creation will be implemented only for supported gang-schedulers (initially Volcano and Coscheduling). Since different schedulers require [the various PodGroups to be created](https://github.com/kubeflow/training-operator/blob/master/pkg/controller.v1/common/job.go#L281).

> And from the example above, how is minMember of 5 derived? 

`minMember` is always equal to `numNodes`. For ML Training all gangs should be alive to execute training. In the future, if we find other use-cases (e.g. HPC), we can discuss it again.

>If this Training API is going t take on the challenge of orchestrating the logic for creation of needed PodGroup (varying by the plugin) then this probably works OK

Yeah, that's correct.

>Otherwise, I'd still place the responsibility on the user for now for creating the pod group, but just allow the schedulerName and other labels to use it in whatever turns into the actual job or podspec.

Btw, this also will be supported, since users can just ignore `PodGroupSpec` parameter, and set the `.spec.schedulerName` field in the PodSpec.
